<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Benchmarking methods for alignment of scRNA-sea and scATAC-seq data</title>
  <meta name="description" content="Benchmarking methods for alignment of scRNA-sea and scATAC-seq data" />
  <meta name="generator" content="bookdown 0.14 and GitBook 2.6.7" />

  <meta property="og:title" content="Benchmarking methods for alignment of scRNA-sea and scATAC-seq data" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Benchmarking methods for alignment of scRNA-sea and scATAC-seq data" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  

<link rel="next" href="results.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />












</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a></li>
<li class="chapter" data-level="2" data-path="results.html"><a href="results.html"><i class="fa fa-check"></i><b>2</b> Results</a><ul>
<li class="chapter" data-level="2.1" data-path="results.html"><a href="results.html#designing-a-bechmark-for-label-transfer-methods-pbmc-dataset"><i class="fa fa-check"></i><b>2.1</b> Designing a bechmark for label transfer methods: PBMC dataset</a></li>
<li class="chapter" data-level="2.2" data-path="results.html"><a href="results.html#optimizing-label-transfer-on-thymus-dataset"><i class="fa fa-check"></i><b>2.2</b> Optimizing label transfer on Thymus dataset</a><ul>
<li class="chapter" data-level="2.2.1" data-path="results.html"><a href="results.html#label-transfer-on-count-gmat"><i class="fa fa-check"></i><b>2.2.1</b> Label transfer on count gmat</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="results.html"><a href="results.html#what-next"><i class="fa fa-check"></i><b>2.3</b> What next?</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="discussion.html"><a href="discussion.html"><i class="fa fa-check"></i><b>3</b> Discussion</a></li>
<li class="chapter" data-level="4" data-path="methods.html"><a href="methods.html"><i class="fa fa-check"></i><b>4</b> Methods</a><ul>
<li class="chapter" data-level="4.1" data-path="methods.html"><a href="methods.html#dataset-details"><i class="fa fa-check"></i><b>4.1</b> Dataset details</a><ul>
<li class="chapter" data-level="4.1.1" data-path="methods.html"><a href="methods.html#pbmc-dataset"><i class="fa fa-check"></i><b>4.1.1</b> PBMC dataset</a></li>
<li class="chapter" data-level="4.1.2" data-path="methods.html"><a href="methods.html#thymus-dataset"><i class="fa fa-check"></i><b>4.1.2</b> Thymus dataset</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="methods.html"><a href="methods.html#data-preprocessing"><i class="fa fa-check"></i><b>4.2</b> Data preprocessing</a></li>
<li class="chapter" data-level="4.3" data-path="methods.html"><a href="methods.html#dimensionality-reduction-and-clustering"><i class="fa fa-check"></i><b>4.3</b> Dimensionality reduction and clustering</a></li>
<li class="chapter" data-level="4.4" data-path="methods.html"><a href="methods.html#scatac-seq-data-embedding"><i class="fa fa-check"></i><b>4.4</b> scATAC-seq data embedding</a></li>
<li class="chapter" data-level="4.5" data-path="methods.html"><a href="methods.html#tested-integration-methods"><i class="fa fa-check"></i><b>4.5</b> Tested integration methods</a><ul>
<li class="chapter" data-level="4.5.1" data-path="methods.html"><a href="methods.html#transfer-label"><i class="fa fa-check"></i><b>4.5.1</b> Label transfer</a></li>
</ul></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Benchmarking methods for alignment of scRNA-sea and scATAC-seq data</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="header">
<h1 class="title">Benchmarking methods for alignment of scRNA-sea and scATAC-seq data</h1>
</div>
<div id="introduction" class="section level1">
<h1><span class="header-section-number">1</span> Introduction</h1>
<p>Following technological advances and discoveries with single-cell transcriptomics, new single-cell sequencing strategies are emerging to profile other molecular layers in thousands of cells, from chromatin accessibility, to expression of surface proteins, to methylation. These methods have already been employed to discover patterns of epigenetic heterogeneity in a variety of tissues [ref]. However, at present these methods produce noisier data and at lower throughput.
Integration of such datasets with more detailed and comprehensively annotated scRNA-seq datasets can allow denoising of biological signals, guide cell type annotation and disentangle causal relationships between biological layers of information and how these co-determine complex phenotypes.</p>
<p>Methods for simoultaneous profiling of biological layers are starting to emerge [refs], but these are mostly low-throughput and labor intensive.
In most cases, molecular profiles will be measured in parallel from cells sampled from the same tissue/cellular population.
Consequently, aligning different datasets requires an at least partial correspondence between profiled features across omics. This is true when integrating different scRNA-seq datasets or scRNA-seq data with spatial transcriptomics. For omic types that measure molecular features that are different than genes, data is usually preprocessed to generate a matrix of gene level features e.g. measuring gene activity from ATAC accessibility peaks [<a href="https://www.ncbi.nlm.nih.gov/pubmed/30078726" class="uri">https://www.ncbi.nlm.nih.gov/pubmed/30078726</a>]. Different integration methods use different inference algorithms for the latent space projection (e.g. canonical correlation analysis, non-negative matrix factorization, variational autoencoders), but all allow the mapping of a gene expression (or other molecular feature) vector x on a latent space vector z, via a vector of feature loadings w. Inspection of loadings can distinguish features that allow alignment across datasets and potentially indicate cell and omic specific contributions to the overall data variation.</p>
<ul>
<li>Finding which method works best</li>
<li>but also comparing different preprocessing strategies</li>
</ul>
<p>Here we implemented an analysis workflow to optimize the performance of methods to transfer labels inferred from a single-cell gene expression dataset to a single-cell accessibility dataset. We compared three published methods based on robustness and performance on a range of metrics. We then applied our workflow to optimize integration of scRNA-seq and scATAC-seq datasets generated from developing human thymus.</p>
</div>
            </section>

          </div>
        </div>
      </div>

<a href="results.html" class="navigation navigation-next navigation-unique" aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
},
"search": false
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
