---
title: "Benchmarking methods for alignment of scRNA-sea and scATAC-seq data"
output: bookdown::pdf_document2
bibliography: references.bib
link-citations: true
---

# Introduction 

Single-cell Assays for Transposase-Accessible Chromatin by sequencing (scATAC-seq) enable profiling of the chromatin accessibility landscape of thousands of single-cells, and identification of *cis-* and *trans-* acting regulatory elements in heterogeneous cell populations [@cusanovichMultiplexSinglecellProfiling2015; @cusanovichSingleCellAtlasVivo2018]. However, common analysis tasks for single-cell RNA-seq (scRNA-seq) datasets, such as clustering and annotation of cell types, are challenging for scATAC-seq data, because of it's extreme sparsity and limited prior knowledge of cell-type specific accessibility patterns.
Integration of accessibility datasets with comprehensively annotated scRNA-seq datasets can allow denoising of biological signal, guide cell type annotation and disentangle causal relationships between biological layers of information and how these co-determine complex phenotypes [@buenrostroIntegratedSingleCellAnalysis2018; @granjaSinglecellMultiomicAnalysis2019]. 

<!-- Methods for simoultaneous profiling of biological layers are starting to emerge [refs], but these are mostly low-throughput and labor intensive. -->
In most cases, molecular profiles will be measured in parallel from cells sampled from the same tissue/cellular population.
Consequently, aligning different datasets requires an at least partial correspondence between profiled features across omics. This is true when integrating different scRNA-seq datasets or scRNA-seq data with spatial transcriptomics. For omic types that measure molecular features that are different than genes, data is usually preprocessed to generate a matrix of gene level features e.g. measuring gene activity from ATAC accessibility peaks [https://www.ncbi.nlm.nih.gov/pubmed/30078726]. Different integration methods use different inference algorithms for the latent space projection (e.g. canonical correlation analysis, non-negative matrix factorization, variational autoencoders), but all allow the mapping of a gene expression (or other molecular feature) vector x on a latent space vector z, via a vector of feature loadings w. Inspection of loadings can distinguish features that allow alignment across datasets and potentially indicate cell and omic specific contributions to the overall data variation.

- Finding which method works best
- but also comparing different preprocessing strategies

Here we implemented an analysis workflow to optimize the performance of methods to transfer labels inferred from a single-cell gene expression dataset to a single-cell accessibility dataset. We compared three published methods based on robustness and performance on a range of metrics. We then applied our workflow to optimize integration of scRNA-seq and scATAC-seq datasets generated from developing human thymus.

# Results

## Designing a benchmark for label transfer methods: PBMC dataset

We started by comparing the performance of three publicly available methods for label transfer: CCA [@stuartComprehensiveIntegrationSingleCell2019], Conos [@barkasJointAnalysisHeterogeneous2019] and Liger [@welchSingleCellMultiomicIntegration2019]. 
We used a publicly available dataset of Peripheral Blood Mononuclear Cells (PBMC) from 10XGenomics. After filtering, this comprises of transcriptomic profiles for 5607 cells and accessibility profiles for 8960 cells. 

<!-- We first evaluate the ability of different methods to transfer cell type annotations derived from scRNA-seq data to scATAC-seq data of the same tissue (See Methods \@ref(transfer-label) for details about label transfer procedure for each method).  -->

Visualizing the predicted labels on the embedding of genome-wide scATAC-seq profiles, we find that for all the methods reconstruct similar clusters, which agree with the global structure of the accessibility data (Fig.\@ref(fig:label-comp)A).
Cell type proportions are in line with those found in the cell population measured with scRNA-seq (Fig.\@ref(fig:label-comp)B).
We next evaluated run time and robustess of methods running label prediction including a growing fraction of scATAC-seq cells. Conos is the fastest method, with speed minimally affected by the number of query cells (Fig.\@ref(fig:label-comp)). Both Conos and CCA were equally robust in their label predictions for subsets of cells, with Adjusted Rand Index compared to labels inferred from the full dataset between 0.6 and 0.8 (Fig.\@ref(fig:label-comp)C). Conversely, Liger showed much higher variability in label prediction.

All methods associate labels with a prediction score, that allows to remove labels with high uncertainty (Fig.\@ref(fig:label-comp)D). We found that different methods score with high confidence similar clusters (Fig.\@ref(fig:label-comp)E). We clustered scRNA-seq cells with increasing resolution (i.e. from bigger few clusters to many smaller clusters) and compared the distribution of prediction scores. We found that Conos predicts labels with higher confidence with larger (low-resolution) clusters (Fig.\@ref(fig:label-comp)F)
<!-- Setting a threshold on the label prediction score allows to remove from downstream analysis cells with a low confidence prediction label, and mark them as unasssigned. We found that at the cutoff of 0.5, suggested by @stuartComprehensiveIntegrationSingleCell2019a, Liger excludes the least amount of cells, while Conos scores the most cells with higher confidence (Fig.\@ref(fig:predict-score)B).  -->

We wanted to quantify if cells that get annotated with the same label also have a similar genome-wide bin accessibility profile. This is crucial to allow reconstruction of consensus chromatin accessibility profiles of clusters/cell types. 
We calculate the fraction of nearest neighbors per cell that share the predicted label. For each prediction, we calculate a null distribution of this fraction by randomly shuffling the assigned labels. We define a KNN purity statistic as the deviation between the null and the true distribution (KS test, see Methods \@ref(knn-purity)). We found that Conos and Liger performed slightly better on this dataset than CCA (Fig.\@ref(fig:label-comp)G).
<!-- Looking at the distributions of KNN scores, we find that Liger maintains the original connectivity structure better, followed by Conos and finally CCA, even if the differences are not that substantial. Also the KNN score is dependent on the assigned cell type. -->


<!-- We checked accessibility of PBMC marker genes in the called cell types. We found that cell populations called with CCA show accessibility of expected marker genes from transcriptomics, for NK cells, monocytes, CD8+ cells, and the platelets even. -->

```{r label-comp, fig.show='hold', fig.cap="Label comp", echo=FALSE}
knitr::include_graphics("output/Fig1.pdf", auto_pdf = TRUE)
```

## Optimizing label transfer on Thymus dataset

We moved on to optimizing label transfer and integration on a dataset capturing cell populations with less distinct molecular signatures. We analyzed a dataset generated from developing thymus at 10 post-conceptional weeks. Libraries for scRNA-seq and scATAC-seq were generated on different samples from the same donor. 
Clustering and annotation of the transcriptomes identified 15 distinct cell populations (Fig|\@ref(fig:opt-thymus)). The main cluster represents T cells undergoing maturation, from the Double Negative (DN) state, to Double Positive (DP) to CD4+ or CD8+ Single Positive (SP) T cells. Other smaller clusters were identified as other components of the thymic microenvironment, such as thymic epithelial cells (TEC) or other immune cell populations (e.g. Dendritic Cells, Macrophages, Innate like cells). Clustering of genome-wide scATAC-seq profiles also partitioned the data with a similar structure, with smaller clusters and a main population divided into subclusters (Fig. ?). 
Repeating the label transfer workflow as for the PBMC dataset we noticed that while all methods consistently assigned the smaller clusters, only with CCA transfered annotations resemble chromatin accessibility clusters, and this is reflected in the KNN purity score (Fig.\@ref(fig:opt-thymus)B). 
We reasoned that the aggregation to gene-level features might be increasing noise in the T cell cluster. To obtain a cell-by-gene matrix that best represents the signal in the genome-wide cell-by-bin matrix, we binarized accessibility by gene so that accessibility is equal to 1 if at least one accessible genomic bin overlaps with the gene body or promoter. We found that binarizing the gene x cell matrix maintains the clustering structure found in the genome-wide accessibility, while this is lost when taking bin counts as accessibility measures of genes (Fig.\@ref(fig:opt-thymus)B). Consequently, using the binary gene matrix for label transfer greatly increased the KNN purity for all methods (Fig.\@ref(fig:opt-thymus)D).

```{r opt-thymus, fig.show='hold', fig.cap="Thymus dataset", echo=FALSE}
knitr::include_graphics("output/Fig2.pdf", auto_pdf = TRUE)
```

## Integration of accessibility and gene expression profiles of T cells
Our comparison of label transfer outcomes indicated that integration with a binary cell-by-gene accessibility matrix and using CCA is the best integration strategy, by our metrics. 
Using these indications, we performed integration af scATAC-seq and scRNA-seq cells focusing on the putative T cell populations in the thymic datasets (Fig.\@ref(fig:tcells)A-B). We modelled the differentiation trajectory of T cells using pseudotime analysis (Fig.\@ref(fig:tcells)C). This ordered the cell populations in both datasets mostly according to the expected order for conventional T cell maturation (Fig.\@ref(fig:tcells)D). We were then able to explore changes in accessibility patterns along the inferred differentiation trajectory. We observed that genome-wide accessibility tends to decrease during T cell maturation (Fig.\@ref(fig:tcells)E), as it is expected for cells moving towards a "primed" state [ref? Maybe @klemmChromatinAccessibilityRegulatory2019 ?]. We then performed motif accessibility analysis with ChromVAR [@schepChromVARInferringTranscriptionfactorassociated2017], which measures enrichment or depletion of accessibility at transcription factor (TF) motifs in scATAC-seq profiles. Among motifs with most variable accessibility in our dataset, we identify motifs for many TFs that have been shown to be involved in T cell maturation (Fig.\@ref(fig:tcells)), such as TCF, RUNX, REL and TBX transcription factors [@hosoyaGlobalDynamicsStagespecific2018, Park et al. 2019, unpublished]. These show distinct changes in accessibility along pseudotime, especially at the very early and final stages of maturation. Motif accessibility showed less variability between DP cell states. Data integration enabled us to compare patterns of TF motif accessibility with patterns of TF expression along pseudotime. We identified instances in which TF expression correlates with motif accessibility, as well as cases in which motif accessibility decreases as TF expression increases (\@ref(fig:tcells)G).

```{r tcells, fig.pos="h", fig.cap="**Integrative analysis of T cell maturation in developint thymus:** Joint visualization of scRNA-seq data with scATAC-seq cells for the T cell clusters, colored by (A) technology, (B) inferred cell type labels and (C) diffusion pseudotime; (D) Distribution of cells along 100 equal-sized pseudotime rank bins, colored by inferred cell type; (E) Distribution of the fraction of accessible genomic bins per cell for each pseudotime rank bin; (F) Heatmap of TF motif deviation for the top 50 most variable TF motifs inferred by chromVAR. Cells are ordered by pseudotime. Values are smoothed with a running average function (step = 30); (G) Smoothed values for log-normalized gene expression counts (top, pink) and TF motif deviation (bottom, cyan) along pseudotime ranks. Values are smoothed with the LOESS function (span = 0.2).", echo=FALSE}
knitr::include_graphics("output/Fig3.pdf", auto_pdf = TRUE)
```

# Discussion
- Methods perform similarly on a dataset with defined clusters
- Does it make a difference how do you reduce ATAC to gene - features? Scoring that takes into account enhancer activity (e.g. Cicero) might be more informative to align differentiating cells.  
- Testing performance on datasets from joint profiling protocols

Conclusions
- CCA seems to outperform the other label transfer methods 
- Binarization of the gene level accessibility data allows assignment of subclusters
- Selecting features based on both datasets is better

# Methods

<!-- ## Dataset details  -->
<!-- Incl. stats about dataset quality (median no. of fragments per cell, perc. of fragments mapping to peaks, median no. of fragments in peaks per cell) -->

### Datasets
**PMBC (10X genomics):** this dataset is composed of Peripheral Blood Mononuclear Cells (PBMCs) from one donor. Raw scRNA-seq counts were downloaded with the Seurat R package [[@satijaSpatialReconstructionSinglecell2015]] (download link: https://www.dropbox.com/s/3f3p5nxrn5b3y4y/pbmc_10k_v3.rds?dl=1). Raw scATAC-seq fragments were downloaded with the SnapATAC R package [@fangFastAccurateClustering2019] (download link: http://renlab.sdsc.edu/r3fang//share/github/PBMC_ATAC_RNA/atac_pbmc_10k_nextgem.snap). A total of 5607 scRNA-seq cells and 8690 scATAC-seq cells were used for analysis after QC. 

**Thymus dataset:** this dataset consists of cells from one fetal thymus at 10 post-conception weeks (Park et al. unpublished, Dominguez-Conde et al. unpublished). A total of 8321 scRNA-seq cells and 5793 scATAC-seq cells were used for analysis after QC. Cell type annotations for the scRNA-seq cells was provided by the authors and based on expression of marker genes.

## Data preprocessing and dimensionality reduction

We preprocessed and normalized scRNA-seq using the standard pipeline from the R package Seurat (v3.1.1). ScATAC-seq reads were aligned and preprocessed using CellRanger (10X genomics). We used the SnapATAC pipeline for quality control and preprocessing  [@FastAccurateClustering2019]. 
This generates genome-wide single-cell accessibility profiles by binning the genome into fixed-size windows (selected bin size: 5 kb) and constructing a cell-by-bin binary matrix, estimating accessibility for each bin.

For dimensionality reduction, we used Principal Component Analysis for scRNA-seq datasets and Latent Semantic Indexing for dimensionality reduction of accessibility matrices, as proposed by @cusanovichMultiplexSinglecellProfiling2015. For data visualization, we used Uniform Manifold Approximation and Projection (UMAP) algorithm [@mcinnesUMAPUniformManifold2018] on low-dimensional data representation.
<!-- ells with high expression of mitochondrial genes were filtered out. Raw counts $c$ were normalized by total cell coverage and converted to $log(c + 1)$ as a variance stabilizing transformation. -->

## Label transfer

**Accessibility in gene-level features:** We generated cell-by-gene count matrices by aggregating bin coverage over the gene bodies and promoters (2kb upstream of transcriptional start site). Count gene matrices were converted to binary gene matrices by substituting counts with 1 if counts > 0.

**Feature selection:** unless otherwise specified, we select genes for label transfer by taking the union of the most highly variable genes in scRNA-seq (using the Seurat function `FindVariableGenes`, with `selection.method = "mvp"`) and the most frequently covered genes in the scATAC-seq datasets (covered in at least 10% of cells). We found that joint feature selection performs significantly better that selection based on the reference dataset only (data not shown).
<!-- Other studies have used only the HVGs from the reference dataset (in most cases the scRNA-seq). We found that these two feature selection strategies gave similar label transfer results and we opted for the union to avoid selecting only genes that have very low coverage in the ATAC-seq data.  -->

**Data integration methods:** details of published integration methods for single-cell data and the rationale behind exclusion from the benchmark are detailed in Table 1.   

| Method       | Reference | Model for embedding | Label/feature propagation | Reason for Excluding |
| ---------------|-------------------|:--------------------------|:-----------------------|:----------------------------|
| Seurat CCA   | @stuartComprehensiveIntegrationSingleCell2019 | Canonical Correlation analysis | mNN pairing | / |
| LIGER | @welchSingleCellMultiomicIntegration2019 | Joint Non-Negative Matrix factorization | KNN graph | / |
| Conos | @barkasJointAnalysisHeterogeneous2019  | Joint PCA | Inter/Intra-dataset edges | / |
| scGen | @lotfollahiScGenPredictsSinglecell2019 | Variational Autoencoder | Decoder | Requires cell type annotation in both datasets |
| totalVI | @gayosoJointModelRNA2019 | Variational inference | Generative model | Requires  multi-omic data from the same single-cells |
| BBKNN | @polanskiBBKNNFastBatch | PCA | Batch balanced graph construction | Bad alignment during testing |
| gimVI | @lopezJointModelUnpaired2019 | Variational inference | Generative model | No implementation for right log-likelihood distribution |
Table: Details of published data integration methods considered for comparison.

<!-- ### Label transfer {#transfer-label} -->
<!-- One of the key tasks for integration methods is to be able to transfer cell type annotations learnt from a reference to a query dataset. This is especially useful if the query is a scATAC-seq dataset, where calling of cell types based on prior knownledge on marker genes is often not possible. Different models are adapted to transfer discrete cell state labels derived from gene expression to cells measured with scATAC-seq. -->

<!-- ##### Seurat CCA -->

<!-- Identified anchor pairs are weighted based on the query cell local neighboorhood (the k nearest anchors) and the anchor score. The obtained reference cells x query cells weight matrix is then multiplied by the matrix of annotation x reference cells, to generate a query cell x annotation matrix. This returns a prediction score for each class for every cell in the query dataset, ranging from 0 to 1 [@stuartComprehensiveIntegrationSingleCell2019a]. -->

<!-- ##### LIGER -->
<!-- While the authors do not describe a method for transferring discrete labels, I adapted their strategy for feature imputation. I build a cross-dataset KNN graph in the aligned factor space, then I assign each query cell to the most abundant label between the k nearest neighbors in the reference dataset (k=30). The prediction score for label $l$ is computed as the fraction of nearest neighbors that have the predicted label. -->
<!-- $$ -->
<!-- score = \frac{count(l)}{k} -->
<!-- $$ -->

<!-- ##### Conos -->
<!-- Label transfer is treated as a general problem of information propagation between vertices of the common graph (detailed in @barkasJointAnalysisHeterogeneous2019). The label score is the label probability updating during the diffusion process. -->

### KNN purity score {#knn-purity}
We construct the k nearest neighbor (KNN) graph of scATAC-seq cells on the LSI reduction matrix (k = 30). For each prediction, we measure the fraction of NNs per cell that have the same predicted label. We do the same after randomly permuting the predicted labels, to estimate a null distribution. The purpose of this step is to avoid giving a high score to a prediction that assigns many cells to just a few clusters. We then compute the Kolmogorov-Smirnov deviation statistic between true and null distribution and define this as KNN purity.
KNN purity score is calculated retaining only labels with prediction score > 0.5. 

### Analysis of T cell maturation
We selected putative T cell clusters based on independent clustering of scATAC-seq and scRNA-seq. We performed co-embedding and label trasfer using the Seurat CCA method [@stuartComprehensiveIntegrationSingleCell2019], with preprocessing as previously described. We imputed gene expression values for scATAC-seq cells and computed diffusion pseudotime [@haghverdiDiffusionPseudotimeRobustly2016] on all cells as implemented in Scanpy (v1.4.4). The cell of origin for the pseudotime algorithm was selected by expression of markers of early DN cells (IGLL1, CD34). [@wolfSCANPYLargescaleSinglecell2018]. For analysis of TF motif enrichment, we used the ChromVAR package [@schepChromVARInferringTranscriptionfactorassociated2017], with default settings.  

### Code availability
All analysis for this report was carried out in R and python. Notebooks and scripts can be accessed at https://github.com/EmmaDann/multiOmic_benchmark.


# Bibliography

<!-- ## Uniform output for all methods  -->
<!-- - Impute transcriptomic data -->
<!-- - **Transfer labels from RNA-seq to ATAC-seq** -->
  

<!-- ## Metrics for comparison of integration models -->

<!-- Problems: finding optimal distance metric for gene accessibility and expression (correlation doesn't work, too sparse).  -->
<!-- (Or more general: how to relate features from different datasets) -->
<!-- Nearest neighbors in PCA? How to make a nearest neighbor graph between modalities  -->
<!-- - MNN cosine normalizes each batch and then calculates distances  -->

<!-- - PCA on integrated space and find genes with high eigen values -->
<!-- - How to denoise the ATAC data?? -->

<!-- 2) Robustness to different fractions of cells in ATAC dataset -->
<!-- 3) Leave-one-out approach for imputed data -->
<!-- 4) **Fraction of unassigned cells** (but how to distinguish unassigned and badly assigned?): prediction score from label transfer  -->
<!-- 5) **Joint clustering: purity of cell type annotation inside a cluster, mixing within the same cluster between different technologies** -->
<!-- 6) **Robustness to parameter picking (e.g. no. of factors)** -->
<!-- 7) Agreement metric defined by Welch et al. 2019: compare KNN graph of single datasets with KNN graph in integrated space, then calculate how many of each cell's NNs in the single dataset graph are also NNs in the integrated graph. Welch et al. compare NN graphs built with different factor models to compare CCA and LIGER performance. I build for all methods KNN graphs from the PCA projection of imputed values. -->
<!-- 8) **Expression not at random of markers after integration:** is the structure that is clear in the RNA maintained after embedding w the ATAC? From idea of JP collaborator -->
<!--     - Find marker genes from RNA only -->
<!--     - Measure non-random expression in RNA only -->
<!-- 9) **pySCENIC on RNA only or integration w ATAC seq** -->

<!-- ## Ideas for less "agnostic" integration -->

<!-- 1) Select only a certain lineage of cells (e.g. that you can align in pseudotime) -->
<!-- 2) Annotation of cell types also in ATAC-seq data (e.g. to use scGen) -->
<!-- 3) Considering enhancer accessibility (matching them to genes??)  -->

<!-- ## Biological interpretation of integration -->

<!-- In Cusanovich et al. 2018 they identify differentially accessible sites & cluster specific accessibility patterns. -->


<!-- ## Other random things -->

<!-- - How does ATAC improve the RNA? Can we detect cell clusters/populations that are highly homogeneous in the RNA -->
<!-- but display significant variability at the accessibility level? But what is variability in super noisy ATAC seq? -->
<!-- - Studying pioneering TFs: temporal relationship between TF expression and accessibility (are they really opening chromatin?)  -->
<!-- Could be done on the time series data @svenssonInterpretableFactorModels2019 -->
<!-- - Clonality of epigenetic modifications ATAC + TCR tracing  -->












