---
title: "Benchmarking methods for alignment of scRNA-sea and scATAC-seq data"
output: github_document
bibliography: bibliography.bib
link-citations: true
---

## Dataset details 
Incl. stats about dataset quality (median no. of fragments per cell, perc. of fragments mapping to peaks, median no. of fragments in peaks per cell)

## Data preprocessing

### ATAC-seq
ATAC-seq data is inherently sparse and noisy, more so than RNA-seq data. In addition, methods for alignment with scRNA-seq data mostly require to reduce accessibility signal to gene-level features. Different papers use different strategies to preprocess ATAC-seq data and featurization (benchmarked by @chenAssessmentComputationalMethods2019a):

- **Seurat pipeline**: raw 10x sequencing reads were preprocessed to call peaks with Cellranger (v ...). Then all counts in peaks are summed up all counts in peaks within gene bodies + 2kb upstream [@welchSingleCellMultiomicIntegration2019a; @stuartComprehensiveIntegrationSingleCell2019a]. 
- **SNAP-ATAC pipeline:** the genome is binned, then a cell x bin binary matrix is constructed. From this binary matrix a gene-level matrix is made [@FastAccurateClusteringa].
- **PCA based on bulk:** some studies [@buenrostroIntegratedSingleCellAnalysis2018] find principal components of differentiation/cell type from bulk ATAC-seq datasets and then project the single-cells on the same space. Arguably, this will lead to miss out on accessibility dynamics of rare cell populations.
- **LSI method:** this procedure starts by making a bin x cell matrix. Then normalization and rescaling are done using term frequency-inverse document frequency transformation (something from text mining), then SVD is performed to obtain a PC x cell matrix. This is used to cluster cells and peaks are called on the clusters. Then the peak x cell matrix is used to do PCA again (I am getting dizzy).



### RNA-seq

Both datasets where preprocessed with std steps: removal of empty/lowQ cells, normalization per cell coverage, 

#### Feature selection
- Highly-variable genes in RNA
- genes that are expressed only in a few cells 

## Tested integration methods 

- **Seurat CCA**:
  - select K?
- **LIGER:**
  - K factors
  - Imputation strategy: projecting ATAC cells on NMF factor space. Testing if this is a valid imputation strategy using scRNA-seq data only 
- **SnapATAC pipeline:** it just wraps CCA alignment
- **scGen:** requires cell type annotation also on the ATAC dataset
- **totalVI:** not applicable as it assumes matching between cells (cite-Seq and RNA-seq from the same single-cells) 
- **BBKNN:** (how do you do an imputation step?)
- **Cusanovich2018:** they figure out a way to transfer cell type annotation labels
- **Conos**


## Uniform output for all methods 
- Impute transcriptomic data

## Metrics for comparison of integration models

Problems: finding optimal distance metric for gene accessibility and expression (correlation doesn't work, too sparse). 
(Or more general: how to relate features from different datasets)
Nearest neighbors in PCA? How to make a nearest neighbor graph between modalities 
- MNN cosine normalizes each batch and then calculates distances 

- PCA on integrated space and find genes with high eigen values
- How to denoise the ATAC data??

2) Robustness to different fractions of cells in ATAC dataset
3) Leave-one-out approach for imputed data
4) **Fraction of unassigned cells** (but how to distinguish unassigned and badly assigned?)
5) **Joint clustering: purity of cell type annotation inside a cluster, mixing within the same cluster between different technologies**
6) **Robustness to parameter picking (e.g. no. of factors)**
7) Agreement metric defined by Welch et al. 2019: compare KNN graph of single datasets with KNN graph in integrated space, then calculate how many of each cell's NNs in the single dataset graph are also NNs in the integrated graph. Welch et al. compare NN graphs built with different factor models to compare CCA and LIGER performance. I build for all methods KNN graphs from the PCA projection of imputed values.
8) **Expression not at random of markers after integration:** is the structure that is clear in the RNA maintained after embedding w the ATAC? From idea of JP collaborator
    - Find marker genes from RNA only
    - Measure non-random expression in RNA only
9) **pySCENIC on RNA only or integration w ATAC seq**

## Ideas for less "agnostic" integration

1) Select only a certain lineage of cells (e.g. that you can align in pseudotime)
2) Annotation of cell types also in ATAC-seq data (e.g. to use scGen)
3) Considering enhancer accessibility (matching them to genes??) 

## Biological interpretation of integration

In Cusanovich et al. 2018 they identify differentially accessible sites & cluster specific accessibility patterns.


## Other random things

- How does ATAC improve the RNA? Can we detect cell clusters/populations that are highly homogeneous in the RNA
but display significant variability at the accessibility level? But what is variability in super noisy ATAC seq?
- Studying pioneering TFs: temporal relationship between TF expression and accessibility (are they really opening chromatin?) 
Could be done on the time series data @svenssonInterpretableFactorModels2019

## Bibliography














