---
title: "Label transfer EDA"
output: html_notebook
---


```{r}
library(Seurat)
library(SnapATAC)
library(tidyverse)
library(DescTools)  # 4 AUC function
library(glue)
library(ggalluvial)  # 4 river plot
library(ggpubr)
source("~/multiOmic_benchmark/utils.R")
source("~/multiOmic_benchmark/KNN_agreement.R")

gg_color_hue <- function(n) {
  hues = seq(15, 375, length = n + 1)
  hcl(h = hues, l = 65, c = 100)[1:n]
}

## Make output directory
outdir <- "~/multiOmic_benchmark/report/output/20191106_labelTransferEDA_PBMC/"
ifelse(!dir.exists(outdir), dir.create(outdir), FALSE)

```


```{r}
# model.cca <- readRDS("~/models/modelCCA_union_hvg_PBMC_SCElist_20191105.RDS")
# model.liger <- readRDS("~/models/modelLiger_union_hvg_PBMC_SCElist_20191105.RDS")
# model.conos <- readRDS("~/models/modelConos_union_hvg_PBMC_SCElist_20191105.RDS")
# 
seu.cca <- readRDS("~/models/labelTransferCCA_union_hvg_PBMC_SCElist_20191105.RDS")
seu.liger <- readRDS("~/models/labelTransferLiger_union_hvg_PBMC_SCElist_20191105.RDS")
seu.conos <- readRDS("~/models/labelTransferConos_union_hvg_PBMC_SCElist_20191105.RDS")

integrate_features <- scan("~/models/intFeatures_union_hvg_2000_F74_SCElist_20191101.txt", what = "")

int.list <- list(CCA=seu.cca, Liger=seu.liger, Conos=seu.conos)

## Make method color palette
method.palette <- brewer_palette_4_values(names(int.list), "Set1")

```

### Embeddings
Visualize label transfer on original ATAC data (embedded SnapATAC bins) and compare with embedding of RNA-seq data
```{r}
## Load original data
orig.ATAC <- readRDS("~/my_data/10X_data/atac_pbmc_10k_nextgem.snapATAC.RDS")
sce.list <- readRDS("~/my_data/10X_data/PBMC_SCElist_20191105.RDS")
orig.RNA <- sce.list$RNA

## Make SeuratObjects
atac.seu <- snapToSeurat(
    obj=orig.ATAC, 
    eigs.dims=1:20, 
    norm=TRUE,
    scale=TRUE
    )
atac.seu <- RenameCells(atac.seu, new.names = orig.ATAC@metaData$barcode)

## Add cell type predictions
getPredictedLabels <- function(seu.int, int.name, id.col="predicted.id", score.col="score", filter_score=0){
  pred.df <- seu.int$ATAC@meta.data[,c(id.col, score.col), drop=F] 
  colnames(pred.df) <- c('predicted.id', "score")
  pred.df <- pred.df %>%
    rownames_to_column("cell") %>%
    mutate(predicted.id = ifelse(score < filter_score, NA, as.character(predicted.id))) %>%
    column_to_rownames("cell")
  rownames(pred.df) <- str_remove(rownames(pred.df), "^ATAC_")
  colnames(pred.df) <- c(str_c("predicted.id", "_", int.name), str_c("score", "_", int.name))
  pred.df
  }

pred.cca <- getPredictedLabels(seu.cca, "CCA", score.col = "prediction.score.max")
pred.liger <- getPredictedLabels(seu.liger, "Liger")
pred.conos <- getPredictedLabels(seu.conos, "Conos")

# pred.cca.union <- getPredictedLabels(seu.cca.union, "CCA.union", score.col = "prediction.score.max")
# cbind(pred.cca, pred.cca.union)


if (all(rownames(pred.conos) == rownames(pred.cca)) & all(rownames(pred.conos) == rownames(pred.liger))) {
  atac.seu <- AddMetaData(atac.seu, metadata = cbind(pred.cca, pred.liger, pred.conos))
} else {
  stop("Non corresponding cell names")
}
```

```{r, fig.height=8, fig.width=18}
## make cell type palette
cell.types <- levels(seu.cca$RNA$annotation)
cell.type.pal <- brewer_palette_4_values(cell.types, palette = "Paired") %>% setNames(cell.types)

## Embed RNA
orig.RNA.seu <- as.Seurat(orig.RNA)
orig.RNA.seu <- FindVariableFeatures(orig.RNA.seu)
orig.RNA.seu <- ScaleData(orig.RNA.seu)
orig.RNA.seu <- RunPCA(orig.RNA.seu)
orig.RNA.seu <- RunUMAP(orig.RNA.seu, dims=1:30)

## Embed ATAC
atac.seu <- RunUMAP(atac.seu, reduction = "SnapATAC", reduction.name = "umap.snap", dims=1:20)

umap_labels_pl <- ggpubr::ggarrange(
    DimPlot(orig.RNA.seu, group.by="annotation", label = TRUE, cols=cell.type.pal, label.size = 3, repel=T) + ggtitle("scRNA-seq") +
       theme( axis.ticks=element_blank(), axis.text = element_blank()),
  plotlist = list(
    DimPlot(atac.seu, reduction = "umap.snap", group.by = "predicted.id_CCA"  , cols=cell.type.pal, label=TRUE, repel=T, label.size = 3) + ggtitle("scATAC-seq - CCA")  + theme( axis.ticks=element_blank(), axis.text = element_blank()),
    DimPlot(atac.seu, reduction = "umap.snap", group.by = "predicted.id_Liger", cols=cell.type.pal, label=TRUE, repel=T, label.size = 3) + ggtitle("scATAC-seq - Liger") + theme( axis.ticks=element_blank(), axis.text = element_blank()),
    DimPlot(atac.seu, reduction = "umap.snap", group.by = "predicted.id_Conos", cols=cell.type.pal, label=TRUE, repel=T, label.size = 3) + ggtitle("scATAC-seq - Conos") + theme( axis.ticks=element_blank(), axis.text = element_blank())
    ),
  common.legend = TRUE, ncol=2, nrow=2
  ) 

umap_labels_pl +
  ggsave(paste0(outdir, "umap_labels.png"), width=8, height = 8)


```
```{r}
umap_labels_pl_long <- ggpubr::ggarrange(
  plotlist = list(
    DimPlot(atac.seu, reduction = "umap.snap", group.by = "predicted.id_CCA"  , cols=cell.type.pal, label=TRUE, repel=T, label.size = 3) + ggtitle("CCA") + theme(axis.ticks=element_blank(), axis.text = element_blank()),
    DimPlot(atac.seu, reduction = "umap.snap", group.by = "predicted.id_Conos", cols=cell.type.pal, label=TRUE, repel=T, label.size = 3) + ggtitle("Conos") + theme(axis.ticks=element_blank(), axis.text = element_blank()),
    DimPlot(atac.seu, reduction = "umap.snap", group.by = "predicted.id_Liger", cols=cell.type.pal, label=TRUE, repel=T, label.size = 3) + ggtitle("Liger") + theme( axis.ticks=element_blank(), axis.text = element_blank()),
    DimPlot(orig.RNA.seu, group.by="annotation", label = TRUE, cols=cell.type.pal, label.size = 3, repel=T) + ggtitle("scRNA-seq") + theme(axis.ticks=element_blank(), axis.text = element_blank())
    ) ,
  common.legend = TRUE, ncol=4, nrow=1
  ) 

umap_labels_pl_long
```


Filter low confidence calls 
```{r, fig.height=8, fig.width=18}
pred.cca.filtered <- getPredictedLabels(seu.cca, "CCA", score.col = "prediction.score.max", filter_score = 0.5)
pred.liger.filtered <- getPredictedLabels(seu.liger, "Liger", filter_score = 0.5)
pred.conos.filtered <- getPredictedLabels(seu.conos, "Conos", filter_score = 0.5)

if (all(rownames(pred.conos) == rownames(pred.cca)) & all(rownames(pred.conos) == rownames(pred.liger))) {
  atac.seu <- AddMetaData(atac.seu, metadata = cbind(pred.cca.filtered, pred.liger.filtered, pred.conos.filtered))
} else {
  stop("Non corresponding cell names")
}

umap_labels_pl <- ggpubr::ggarrange(
  plotlist = list(
    DimPlot(atac.seu, reduction = "umap.snap", group.by = "predicted.id_CCA", cols=cell.type.pal) + 
      scale_color_manual(values = cell.type.pal, na.value="grey80") +
      ggtitle("CCA"),
    DimPlot(atac.seu, reduction = "umap.snap", group.by = "predicted.id_Liger", cols=cell.type.pal) + 
      scale_color_manual(values = cell.type.pal, na.value="grey80") + ggtitle("Liger"),
    DimPlot(atac.seu, reduction = "umap.snap", group.by = "predicted.id_Conos", cols=cell.type.pal) + 
      scale_color_manual(values = cell.type.pal, na.value="grey80") + ggtitle("Conos"),
    DimPlot(orig.RNA.seu, reduction = "umap", group.by = "annotation", cols=cell.type.pal) + ggtitle("RNA") +
      scale_color_manual(values = cell.type.pal, na.value="grey80")
  ),
  common.legend = TRUE, ncol=4, nrow=1
) +
  ggsave(paste0(outdir, "umap_labels_filtered.png"), width=16, height = 6)

```

<!-- ```{r} -->
<!-- pl <-     DimPlot(atac.seu, reduction = "umap.snap", group.by = "predicted.id_Liger", cols=cell.type.pal, label=TRUE, repel=TRUE) + ggtitle("Liger") -->
<!-- plotly::ggplotly(pl) -->
<!-- ``` -->

## Prediction score
Quantifies the uncertainty of the prediction. Calculated differently for every method, but used to define which cells are "unassigned".


```{r, fig.height=7, fig.width=9}
orig.composition <- orig.RNA$annotation
orig.frac <- table(orig.composition)/length(orig.composition)

orig.frac.df <- data.frame(orig.frac) %>%
  dplyr::rename(predicted.id=orig.composition, frac.label=Freq) %>%
  mutate(method="original.RNA")

score_cols <- str_subset(colnames(atac.seu@meta.data), 'score_')
label_cols <- str_subset(colnames(atac.seu@meta.data), 'predicted.id_')

pred.labels.df <- imap(list(CCA=pred.cca, Liger=pred.liger, Conos=pred.conos), ~ 
      rownames_to_column(.x, "cell") %>%
      rename_all(funs(str_remove(., str_c("_",.y)))) %>%
      mutate(method=.y)
    ) %>%
  purrr::reduce(bind_rows) %>%
  mutate(score=ifelse(is.na(score), 0, score))

predict_score_hist <- 
  pred.labels.df %>%
  ggplot(aes(score, fill=method)) +
  geom_histogram(position="identity", alpha=0.8, bins=40) +
  facet_grid(. ~ method) +
  scale_fill_brewer(palette="Set1") +
  xlab("Label prediction score") +
  theme_bw(base_size = 16) +
  theme(legend.position = "top")

umap.score.cca <- FeaturePlot(atac.seu, reduction = "umap.snap", feature = "score_CCA"  , coord.fixed = TRUE) + 
  scale_color_viridis_c(name="Prediction\nscore") + ggtitle("CCA") 
umap.score.conos <- FeaturePlot(atac.seu, reduction = "umap.snap", feature = "score_Conos", coord.fixed = TRUE) + 
  scale_color_viridis_c() + ggtitle("Conos") + guides(color="none")
umap.score.liger <- FeaturePlot(atac.seu, reduction = "umap.snap", feature = "score_Liger", coord.fixed = TRUE) + 
  scale_color_viridis_c() + ggtitle("Liger") + guides(color="none")

predict_score_umaps <- ggpubr::ggarrange(
  plotlist = list(umap.score.cca + guides(color="none"), umap.score.conos, umap.score.liger) %>%
    map(~ .x + theme(axis.ticks=element_blank(), axis.text = element_blank())),
  ncol=3, nrow=1, legend = "bottom", common.legend = T
)

predict_score_umaps
leg <- get_legend(umap.score.cca)

score_pl <- plot_grid(plotlist = list(predict_score_hist + theme(legend.position = "none"), NULL,predict_score_umaps, leg), nrow=2, ncol=2, rel_heights = c(1,1.2), rel_widths = c(2,0.5)) 

score_pl +
  ggsave(paste0(outdir, "prediction_score_distribution.png"), height = 6, width = 9)
```


## Cell type composition
Compare cell type fractions (w uncertainty)
<!-- Compare cell type fractions (w uncertainty) -->
<!-- ```{r, fig.height=7, fig.width=6} -->

<!-- pred.labels.df %>% -->
<!--   group_by(method) %>% -->
<!--   drop_na() %>% -->
<!--   mutate(tot.cells=n()) %>% -->
<!--   ungroup() %>% -->
<!--   group_by(method, predicted.id) %>% -->
<!--   summarise(tot.label = n(), tot.cells = max(tot.cells), mean.score=mean(score)) %>% -->
<!--   mutate(frac.label=tot.label/tot.cells) %>% -->
<!--   # bind_rows(orig.frac.df) %>% -->
<!--   ggplot(aes(method, predicted.id)) + -->
<!--   geom_point(aes(color=mean.score, size=frac.label)) + -->
<!--   scale_color_viridis_c(name="Mean prediction\nscore") + -->
<!--   scale_shape(name="Fraction of cells") + -->
<!--   # scale_color_gradient(low ="grey", high="red") + -->
<!--   theme_classic(base_size = 16) + -->
<!--   ggsave(paste0(outdir, "cell_type_composition.png"), width=6, height = 6) -->


<!-- ``` -->

```{r, fig.width=14, fig.height=7}
orig.rank.df <- orig.frac.df %>% 
  mutate(orig.rank=dense_rank(frac.label)) %>%
  select(orig.rank, predicted.id) %>%
  distinct() %>%
  arrange(orig.rank) %>%
  column_to_rownames("predicted.id") 

ct_comp_pl <- pred.labels.df %>%
  group_by(method) %>%
  drop_na() %>%
  mutate(tot.cells=n()) %>%
  ungroup() %>%
  group_by(method, predicted.id) %>%
  summarise(tot.label = n(), tot.cells = max(tot.cells), mean.score=mean(score)) %>%
  mutate(frac.label=tot.label/tot.cells) %>%
  bind_rows(orig.frac.df) %>%
  mutate(orig.rank = orig.rank.df[predicted.id,]) %>%
  mutate(predicted.id=factor(predicted.id, levels=rownames(orig.rank.df)))%>%
  ungroup() %>%
  mutate(method=ifelse(method=="original.RNA", "scRNA-seq", method)) %>%
  ggplot(aes(predicted.id, frac.label, fill=predicted.id, color=predicted.id)) +
  geom_point(size=2) +
  geom_col(width=0.1) +
  coord_flip() +
  facet_wrap(method~., nrow=1, ncol=4, scales="free_x") +
  scale_color_manual(values=cell.type.pal) +
  scale_fill_manual(values=cell.type.pal) +
  # scale_color_viridis_c(name="Prediction\nscore") +
  # scale_fill_viridis_c(name="Prediction\nscore") +
  ylab("Fraction of cells") + xlab("") +
  theme_bw(base_size = 16) +
  guides(fill="none", color="none")

ct_comp_pl +
  ggsave(paste0(outdir, "cell_type_composition_bars.png"), width = 9, height = 4)
```

<!-- Does the uncertainty depend on the size of the cluster? -->
<!-- ```{r, fig.width=14, fig.height=5} -->

<!-- pred.labels.df %>% -->
<!--   group_by(method) %>% -->
<!--   drop_na() %>% -->
<!--   mutate(tot.cells=n()) %>% -->
<!--   ungroup() %>% -->
<!--   group_by(method, predicted.id) %>% -->
<!--   summarise(tot.label = n(), tot.cells = max(tot.cells), mean.score=median(score), sd.score=mad(score)) %>% -->
<!--   mutate(frac.label=tot.label/tot.cells) %>% -->
<!--   # bind_rows(orig.frac.df) %>% -->
<!--   ggplot(aes(frac.label, mean.score, color=method)) + -->
<!--   geom_point(size=2) + -->
<!--   geom_errorbar(aes(ymin=mean.score-sd.score, ymax=mean.score+sd.score), alpha=0.6) + -->
<!--   scale_color_brewer(palette="Set1") + -->
<!--   # geom_smooth(method = "loess", span=1.2) + -->
<!--   facet_grid(. ~ method) + -->
<!--   theme_bw(base_size = 16) + -->
<!--   stat_cor(label.x = 0.2, label.y=0.25, color="black", size=5) + -->


<!-- ``` -->

### Agreement with unsupervised clustering of ATAC data
Calculate which fractions of NNs in bin based graph of ATAC cells have the same annotation
```{r}
k = 30
atac.seu <- FindNeighbors(atac.seu, assay = "ATAC", reduction = "SnapATAC", dims = 1:20, k.param = k)

atac.nn.list <- getNNlist(atac.seu)

knn.score.CCA <- test.knn(atac.nn.list, setNames(pred.cca$predicted.id_CCA, rownames(pred.cca)))
knn.score.conos <- test.knn(atac.nn.list, setNames(pred.conos$predicted.id_Conos, rownames(pred.conos)))
knn.score.liger <- test.knn(atac.nn.list, setNames(pred.liger$predicted.id_Liger, rownames(pred.liger)))

knn_score_df <-
  list(CCA=knn.score.CCA, conos=knn.score.conos, liger=knn.score.liger) %>%
  imap( ~ data.frame(KNN_score = .x$KNN_score, D=.x$D, p.val=.x$p.val, method=.y)) %>%
  # imap( ~ data.frame(KNN_score = .x$KNN_score, cell= names(.x$KNN_score), D=.x$D, p.val=.x$p.val, method=.y)) %>%
  purrr::reduce(bind_rows) %>%
  dplyr::mutate(KNN_score=ifelse(is.na(KNN_score), 0, KNN_score)) %>%
  mutate(data="True")
knn_score_null_df <-
  list(CCA=knn.score.CCA, conos=knn.score.conos, liger=knn.score.liger) %>%
  imap( ~ data.frame(KNN_score = .x$null, D=.x$D, p.val=.x$p.val, method=.y)) %>%
  # imap( ~ data.frame(KNN_score = .x$KNN_score, cell= names(.x$KNN_score), D=.x$D, p.val=.x$p.val, method=.y)) %>%
  purrr::reduce(bind_rows) %>%
  dplyr::mutate(KNN_score=ifelse(is.na(KNN_score), 0, KNN_score)) %>%
  mutate(data="Null")


knn_score_pl <- bind_rows(knn_score_df, knn_score_null_df) %>%
  ggplot(aes(KNN_score, color=method)) +
  stat_ecdf( aes(alpha=data), size=1) +
  # stat_ecdf(data=. %>% filter(data=="true"), size=1) +
  facet_grid(.~method) +
  scale_alpha_discrete( range=c(0.5,1), name="") +
  scale_color_brewer(palette = "Set1") +
  guides(color='none') +
  geom_text(data=. %>% distinct(method, D, p.val), 
            x=1, y=0.05, hjust=1,
            aes(label=glue("KNN score = {round(D, 3)}\np.value: {p.val}"), y=c(0.90, 0.95, 1))) +
  theme_bw(base_size = 16) +
  ylab("ECDF") + xlab("Fraction of KNNs with shared label") 

knn_score_pl +
  ggsave(paste(outdir,"KNN_score_ecdf_unionHVG.png"), height = 4, width=9)
```
```{r}
KNN_purity_pl <-
  bind_rows(knn_score_df, knn_score_null_df) %>%
  group_by(method) %>%
  summarise(D=max(D)) %>%
  mutate(method=c(CCA='CCA', conos="Conos", liger="Liger")[method]) %>%
  ggplot(aes(method, D, fill=method, color=method)) +
  geom_col(width=0.5) +
  coord_flip() +
  scale_color_brewer(palette = "Set1") +
  scale_fill_brewer(palette = "Set1") +
  theme_bw(base_size = 16) +
  ylim(0,1) +
  ylab("KNN purity") + xlab("") +
  theme(legend.position = "top")
KNN_purity_pl
KNN_purity_pl + ggsave(paste0(outdir, "KNN_purity_bars.png"), height = 3, width=6) 
```



```{r, fig.height=5, fig.width=13}
compareCluster <- function(clust){
  plotlist <- map(list("CCA", "Liger", "Conos"), 
                  ~ DimPlotCluster(atac.seu, annotation_col = glue('predicted.id_{.x}'), cluster=clust, label=glue('{clust} - {.x}'))
                  )
  ggarrange(plotlist = plotlist, nrow=1)
  }

compareCluster("CD14+ Monocytes") +
  ggsave(paste0(outdir, 'CD14Monocytes_cluster.png'), width=14, height = 4)
compareCluster("CD8 Naive")
```

Assemble figure
```{r, fig.height=9, fig.width=12}
fig.1 <- plot_grid(
  # umap_labels_pl,
  plotlist = list(predict_score_hist + theme(legend.position = "none"), 
                  predict_score_umaps, 
                  knn_score_pl + guides(alpha="none")
                  ),
  rel_heights = c(1, 1.1, 1.3), rel_widths = c(2,1), nrow=3, align = "v", axis = "l"
  )

fig.1.leg <- plot_grid(fig.1, plot_grid(plotlist = list(NULL, leg, NULL), nrow=3, ncol=1), rel_widths = c(3,1), align = "h", axis = "top")

fig.2 <- plot_grid(
  umap_labels_pl_long,
  ct_comp_pl + theme(legend.position = "none"),
  nrow=2, ncol=1, align = "v", axis = "l"
  )
fig.2 + ggsave(paste0(outdir, "fig_PBMC_celltypes.png"), width=13, height = 8)

plot_grid( fig.2, fig.1.leg, nrow=2, align="v", axis = "l") +
  ggsave(paste0(outdir, "fig_PBMC.png"), width=11, height = 17)
```


<!-- ciaone -->
<!-- ```{r, fig.height=8, fig.width=7, message=FALSE} -->
<!-- # plot_KNNecdf <- function(cluster){ -->
<!-- #   full_join(pred.labels.df, knn_score_df) %>% -->
<!-- #     filter(predicted.id==cluster) %>% -->
<!-- #     ggplot(aes(KNN_score, color=method)) + -->
<!-- #     stat_ecdf(size=0.8) + -->
<!-- #     facet_wrap("predicted.id") + -->
<!-- #     xlim(0,1) + ylim(0,1) + -->
<!-- #     coord_fixed() + -->
<!-- #     scale_color_brewer(palette = "Set1") + -->
<!-- #     theme_bw(base_size = 16) + -->
<!-- #     theme(legend.position = "top") -->
<!-- # } -->
<!-- # -->
<!-- DimPlotCluster <- function(annotation_col, cluster, label){ -->
<!--   highlight = which(atac.seu@meta.data[,annotation_col]==cluster) -->
<!--   DimPlot(atac.seu, reduction = "umap.snap",cells.highlight = highlight, cols.highlight = "red", pt.size = 0.02, sizes.highlight = 0.1) + -->
<!--     guides(color="none") + -->
<!--     ggtitle(label = label) -->
<!--   } -->
<!-- # -->
<!-- # UMAPs_cluster <- function(cluster){ -->
<!-- #   ggarrange(plotlist=imap(list(CCA="predicted.id_CCA", Conos="predicted.id_Conos", Liger="predicted.id_Liger"), ~ DimPlotCluster(.x, cluster, label = .y )), ncol=3, nrow=1) %>% annotate_figure(cluster) -->
<!-- # } -->
<!-- # -->
<!-- # map(cell.types, ~ ggarrange(plot_KNNecdf(.x), UMAPs_cluster(.x), nrow = 2, heights = c(1,0.8))) -->
<!-- # -->

<!-- ``` -->

<!-- #### Which cells are inconsistently aligned? -->
<!-- ```{r, fig.width=14, fig.height=10} -->
<!-- pred.labels.df %>% -->
<!--   select(method, predicted.id, cell) %>% -->
<!--   mutate(predicted.id=ifelse(is.na(predicted.id), "none", predicted.id)) %>% -->
<!--   ggplot(aes(x=method, stratum=predicted.id, alluvium=cell, fill=predicted.id, label=predicted.id)) + -->
<!--   geom_flow() + -->
<!--   geom_stratum(color=NA) + -->
<!--   geom_text(stat="stratum") + -->
<!--   theme_bw(base_size = 16) -->
<!-- ``` -->


## Closer look at Liger
```{r}
library(liger)

plotByDatasetAndCluster(model.liger$model,   clusters=c(set_names(as.character(orig.RNA$annotation), colnames(orig.RNA)), set_names(as.character(pred.liger[,'predicted.id_Liger']), rownames(pred.liger))))
```

## Accessibility of markers for cell type
Markers taken from scanpy tutorial
```{r, fig.height=10, fig.width=10, warning=FALSE, message=FALSE}
# pbmc.markers <- c("CD79A", "MS4A1", "CD8A", "CD8B", "LYZ")
pbmc.markers <- list(`T-cell`=c("CD3D"), `B-cell`=c("CD79A", "MS2A1"), `T-cell CD8+`=c("CD8A", "CD8B"), 
                     `T-memory`=c("CAPG", "GABARAPL2", "CRIP1"),
                     NK=c("GNLY", "NKG7"), Myeloid=c("CST3", "LYZ"), Monocytes=c("FCG3A", "C1QA", "PSAP"), 
                     Dendritic=c("FCER1A", "LST1"), Platelets=c("PF4"))
pbmc.markers.df <- imap(pbmc.markers, ~ data.frame(gene=.x, cell.type.class=.y)) %>% 
  purrr::reduce(bind_rows)

marker.access.df <- atac.seu@assays$ACTIVITY@data[intersect(unlist(pbmc.markers), rownames(atac.seu@assays$ACTIVITY)),] %>%
  as.matrix() %>%
  reshape2::melt(varnames=c("gene", "cell"), value.name="log.counts") %>%
  full_join(rownames_to_column(atac.seu@meta.data[, label_cols], "cell")) %>%
  full_join(pbmc.markers.df) %>%
  pivot_longer(cols=label_cols, names_to = "method", values_to = "predicted.id") %>%
  dplyr::mutate(method=str_remove(method,".+_")) %>%
  filter(method %in% c("CCA", "Liger", "Conos")) %>%
  group_by(method, predicted.id, gene) %>%
  dplyr::mutate(frac.cells=sum(log.counts > 0)/n()) 

  
markers_pl <- ggplot(marker.access.df, aes( gene, predicted.id)) +
  geom_point(aes(size=frac.cells, color=frac.cells)) +
  facet_grid(method~cell.type.class, space="free", scales="free_x") +
  scale_color_gradient(high="red", low="white") +
  cowplot::theme_cowplot(font_size = 16) +
  theme(axis.text.x = element_text(angle=90, hjust=1, vjust=0.5),
        strip.text.x = element_text(angle=45)) 

markers_pl +
  ggsave(paste0(outdir, "PBMC_markers_accessibility.png"), height = 10, width = 12)
```



## Compare feature selection strategy
```{r}
model.cca.union <- readRDS("~/models/modelCCA_union_hvg_PBMC_SCElist_20191105.RDS")
seu.cca.union <- readRDS("~/models/labelTransferCCA_union_hvg_PBMC_SCElist_20191105.RDS")
seu.liger.union <- readRDS("~/models/labelTransferLiger_union_hvg_PBMC_SCElist_20191105.RDS")
seu.conos.union <- readRDS("~/models/labelTransferConos_union_hvg_PBMC_SCElist_20191105.RDS")

integrate_features <- model.cca.union$model@anchor.features

int.list <- list(CCA=seu.cca, Liger=seu.liger, Conos=seu.conos)

## Add to ATAC object meta.data
pred.cca.union <- getPredictedLabels(seu.cca.union, "CCA_union", score.col = "prediction.score.max")
pred.liger.union<- getPredictedLabels(seu.liger.union, "Liger_union")
pred.conos.union<- getPredictedLabels(seu.conos.union, "Conos_union")

# pred.cca.union <- getPredictedLabels(seu.cca.union, "CCA.union", score.col = "prediction.score.max")
# cbind(pred.cca, pred.cca.union)


if (all(rownames(pred.conos) == rownames(pred.cca)) & all(rownames(pred.conos) == rownames(pred.liger))) {
  atac.seu <- AddMetaData(atac.seu, metadata = cbind(pred.cca.union, pred.liger.union, pred.conos.union))
} else {
  stop("Non corresponding cell names")
}

```

```{r, fig.width=10, fig.height=14}
ggpubr::ggarrange(
  plotlist = list(
    DimPlot(atac.seu, reduction = "umap.snap", group.by = "predicted.id_CCA"  , cols=cell.type.pal, label=TRUE, repel=TRUE) + ggtitle("CCA"),     DimPlot(atac.seu, reduction = "umap.snap", group.by = "predicted.id_CCA_union"  , cols=cell.type.pal, label=TRUE, repel=TRUE) + ggtitle("CCA union"),
    DimPlot(atac.seu, reduction = "umap.snap", group.by = "predicted.id_Liger", cols=cell.type.pal, label=TRUE, repel=TRUE) + ggtitle("Liger"),
        DimPlot(atac.seu, reduction = "umap.snap", group.by = "predicted.id_Liger_union", cols=cell.type.pal, label=TRUE, repel=TRUE) + ggtitle("Liger_union"),
    DimPlot(atac.seu, reduction = "umap.snap", group.by = "predicted.id_Conos", cols=cell.type.pal, label=TRUE, repel=TRUE) + ggtitle("Conos"),
    DimPlot(atac.seu, reduction = "umap.snap", group.by = "predicted.id_Conos_union", cols=cell.type.pal, label=TRUE, repel=TRUE) + ggtitle("Conos union")
      ),
  common.legend = TRUE, ncol=2, nrow=3
)
```

```{r, fig.width=16, fig.height=8}
ggpubr::ggarrange(
  plotlist = list(
    FeaturePlot(atac.seu, reduction = "umap.snap", feature = "score_CCA_union"  , coord.fixed = TRUE) + ggtitle("CCA"),
    FeaturePlot(atac.seu, reduction = "umap.snap", feature = "score_Liger_union", coord.fixed = TRUE) + ggtitle("Liger"),
    FeaturePlot(atac.seu, reduction = "umap.snap", feature = "score_Conos_union", coord.fixed = TRUE) + ggtitle("Conos")
  ),
  common.legend = TRUE, ncol=3, nrow=1
) 
```

```{r}
orig.composition <- orig.RNA$annotation
orig.frac <- table(orig.composition)/length(orig.composition)

orig.frac.df <- data.frame(orig.frac) %>%
  dplyr::rename(predicted.id=orig.composition, frac.label=Freq) %>%
  mutate(method="original.RNA")

score_cols_union <- str_subset(colnames(atac.seu@meta.data), 'score_.+_union')
label_cols_union <- str_subset(colnames(atac.seu@meta.data), 'predicted.id_.+union')

pred.labels.union.df <- imap(list(CCA=pred.cca.union, Liger=pred.liger.union, Conos=pred.conos.union), ~ 
      rownames_to_column(.x, "cell") %>%
      rename_all(funs(str_remove(., str_c("_",.y)))) %>%
      mutate(method=.y)
    ) %>%
  purrr::reduce(bind_rows) %>%
  mutate(score=ifelse(is.na(score_union), 0, score_union))

predict_score_hist <- 
  pred.labels.union.df %>%
  ggplot(aes(score_union, fill=method)) +
  geom_histogram(position="identity", alpha=0.8, bins=40) +
  facet_grid(method ~.) +
  scale_fill_brewer(palette="Set1") +
  xlab("Label prediction score") +
  theme_bw(base_size = 16) +
  theme(legend.position = "top")

cutoffs <- seq(0,1,0.05)
predict_score_cumedist <-
  pred.labels.union.df %>%
  group_by(method) %>%
  mutate(bins=cut(score_union, breaks = cutoffs)) %>%
  mutate(score=as.numeric(str_remove_all(as.character(bins), ".+,|]"))) %>%
  ggplot(aes(score, color=method)) +
  stat_ecdf(size=0.8, alpha=0.7) +
  scale_color_brewer(palette = "Set1") +
  ylab("Fraction of unassigned cells") +
  xlab("Prediction score cutoff") +
  theme_bw(base_size = 16) +
  xlim(0,1) +
  coord_fixed() +
  guides(color="none") 

ggarrange(predict_score_hist, predict_score_cumedist, common.legend = TRUE, widths = c(0.8, 1.2),
          labels=c("A", "B")) +
  ggsave(paste0(outdir, "prediction_score_distribution_unionHVG.png"), height = 6, width = 10)
```
```{r, fig.width=10, fig.height=10}

full_join(
  pred.labels.df,
  select(pred.labels.union.df, cell, predicted.id_union, score_union, method),
  by=c("cell", "method")
  ) %>%
  group_by(method, predicted.id) %>%
  mutate(n_pred=n()) %>%
  ungroup() %>%
  group_by(method, predicted.id, predicted.id_union) %>%
  summarise(n=n(), n_pred=max(n_pred)) %>%
  mutate(frac=n/n_pred) %>%
  ggplot(aes(predicted.id, predicted.id_union)) +
  geom_tile(aes(fill=frac)) +
  facet_wrap(method~., nrow=3, ncol=1) +
  coord_fixed() +
  scale_fill_gradient(low="white", high="red") +
  xlab("Feat. selection: union HVG") + ylab("Feat. selection: union HVG") +
  theme_cowplot(font_size = 16) +
  theme(axis.text.x = element_text(angle=45, hjust=1)) +
  ggsave(paste0(outdir, "unionVSreference.png"), height = 12, width=10)
```
```{r}
k = 50
atac.seu <- FindNeighbors(atac.seu, assay = "ATAC", reduction = "SnapATAC", dims = 1:15, k.param = k)

atac.nn.list <- getNNlist(atac.seu)

calculate_KNN_agreement <- function(pred.cca, pred.conos, pred.liger, k=50){
  score.CCA <- imap_dbl(atac.nn.list, ~ sum(pred.cca[.x,1] == pred.cca[.y,1])/k) %>% setNames(names(atac.nn.list))
  score.Conos <- imap_dbl(atac.nn.list, ~ sum(pred.conos[.x,1] == pred.conos[.y,1])/k) %>% setNames(names(atac.nn.list))
  score.Liger <- imap_dbl(atac.nn.list, ~ sum(pred.liger[.x,1] == pred.liger[.y,1])/k) %>% setNames(names(atac.nn.list))
  
  knn_score_df <-
    as.data.frame(cbind(score.Conos, score.Liger, score.CCA)) %>%
    rownames_to_column("cell") %>%
    pivot_longer(cols=str_subset(colnames(.), "score"), names_to = "method", values_to = "KNN_score") %>%
    dplyr::mutate(KNN_score=ifelse(is.na(KNN_score), 0, KNN_score),
                  method=str_remove(method, "score."))
  
  quants = seq(0,1, by = 0.05)
  AUECDF_knn_score <- knn_score_df %>%
    split(.$method) %>%
    map_dbl( ~ .x %>%
        arrange(KNN_score) %>% 
        {ecdf(.$KNN_score)(quants)} %>% AUC(quants,.)
      )
  list(knn_score_df, AUECDF_knn_score)  
}
  
knn_agreement <- calculate_KNN_agreement(pred.cca.union, pred.conos.union, pred.liger.union, k=50)
knn_agreement[[1]] %>%
  mutate(AUC=knn_agreement[[2]][method]) %>%
  ggplot(aes(KNN_score, color=method, fill=method)) +
  stat_ecdf(size=1) +
  scale_color_brewer(palette = "Set1") +
  geom_text(data=. %>% group_by(method) %>% summarise(AUC=max(AUC)), 
            x=0, hjust=0, size=5,
            aes(label=glue("AUC = {round(AUC, 3)}"), y=c(0.90, 0.95, 1))) +
  theme_bw(base_size = 16) +
  ylab("ECDF") +
  ggsave(paste(outdir,"KNN_score_ecdf_unionHVG.png"), height = 4, width=6)
```

### Aggregate clusters
```{r}
clust = "NK cell"

cluster.bmat <- orig.ATAC@bmat[which(pred.cca$predicted.id_CCA==clust),]
clust.sums <- as.matrix(cluster.bmat) %>% colSums()

plot(clust.sums[1:2000])
```


### Thoughts
- Conos scores a lot of cells with high confidence, but fails to assign cells to difficult clusters 
- CCA resembles the composition of the RNA data better, but curious that the other methods identify way more 










