---
title: "test_SeuratCCA"
output: html_notebook
---

From [vignette](https://satijalab.org/seurat/v3.0/atacseq_integration_vignette.html)

```{r}
library(tidyverse)
library(Seurat)
source('./ATAC_preprocess/featurizeATAC.R')
```

Load data
```{r}
## ATAC
peak_file <- "../my_data/10X_data/atac_v1_pbmc_10k_filtered_peak_bc_matrix.h5"
annotation_file <- "~/annotations/Homo_sapiens.GRCh38.86.gtf"
## RNA
pbmc.rna <- readRDS("../my_data/10X_data/pbmc_10k_v3.rds")
pbmc.rna$tech <- "rna"
```

1. Make Gene Activity Matrix
```{r}
activity.matrix <- CreateGeneActivityMatrix(peak.matrix = peaks, annotation.file = "../annotations/Homo_sapiens.GRCh37.82.gtf", 
    seq.levels = c(1:22, "X", "Y"), upstream = 2000, verbose = TRUE)
```

2. Object setup
```{r}
pbmc.atac <- CreateSeuratObject(counts = peaks, assay = "ATAC", project = "10x_ATAC")
pbmc.atac[["ACTIVITY"]] <- CreateAssayObject(counts = activity.matrix)
# meta <- read.table("../data/atac_v1_pbmc_10k_singlecell.csv", sep = ",", header = TRUE, row.names = 1, 
    # stringsAsFactors = FALSE)
# meta <- meta[colnames(pbmc.atac), ]
# pbmc.atac <- AddMetaData(pbmc.atac, metadata = meta)
pbmc.atac <- subset(pbmc.atac, subset = nCount_ATAC > 5000)
pbmc.atac$tech <- "atac"
```

3. Data preprocessing 
```{r}
DefaultAssay(pbmc.atac) <- "ACTIVITY"
pbmc.atac <- FindVariableFeatures(pbmc.atac)
pbmc.atac <- NormalizeData(pbmc.atac)
pbmc.atac <- ScaleData(pbmc.atac)
```
4. Viz
```{r}
VariableFeatures(pbmc.atac) <- names(which(Matrix::rowSums(pbmc.atac) > 400))
pbmc.atac <- RunPCA(pbmc.atac, assay = 'ACTIVITY', features = VariableFeatures(pbmc.atac))
pbmc.atac <- RunUMAP(pbmc.atac, reduction = "pca", dims = 1:50)

p1 <- DimPlot(pbmc.atac, reduction = "umap") + NoLegend() + ggtitle("scATAC-seq")
p2 <- DimPlot(pbmc.rna, group.by = "celltype", label = TRUE, repel = TRUE) + NoLegend() + ggtitle("scRNA-seq")
CombinePlots(plots = list(p1, p2))
```
5. Transfer anchors
```{r}
transfer.anchors <- FindTransferAnchors(reference = pbmc.rna, query = pbmc.atac, features = VariableFeatures(object = pbmc.rna), 
    reference.assay = "RNA", query.assay = "ACTIVITY", reduction = "cca")

transfer.anchors@object.list[[1]]@reductions$cca.l2@cell.embeddings[,1:10] %>%
  as_tibble(rownames = "cell") %>%
  mutate(assay=ifelse(str_detect(cell, "query"), "ATAC", "RNA")) %>%
  ggplot(aes(L2CC_1, L2CC_3, color=assay)) +
  geom_point(size=0.5, alpha=0.2)
```
```{r}
integration.anchors <- FindIntegrationAnchors(object.list = list(rna=pbmc.rna, atac=pbmc.atac), anchor.features = VariableFeatures(object = pbmc.rna), 
    reference = 1, reduction = "cca")

```

```{r}
pbmc.integrated <- IntegrateData(integration.anchors)
```


```{r}
pbmc.integrated <- ScaleData(pbmc.integrated)
pbmc.integrated <- RunPCA(object = pbmc.integrated, verbose = FALSE)
pbmc.integrated <- RunUMAP(object = pbmc.integrated, dims = 1:30)
```
```{r}
DimPlot(pbmc.integrated, reduction = "umap", cols="Method")
```

