---
title: "ATAC preprocessing with Seurat pipeline "
output: html_notebook
---

```{r}
suppressPackageStartupMessages({
library(Seurat)
library(tidyverse)
library(AnnotationHub)
library(Organism.dplyr)
library(pbapply)
library(future)
library(reshape2)
})
```

Load peaks and create feature matrix. Annotation file downloaded from `ftp://ftp.ensembl.org/pub/release-86/gtf/homo_sapiens/Homo_sapiens.GRCh38.86.gtf.gz`. I transformed the gtf annotation file to comply with the UCSC annotation (starting with 'chr') (running `cat annotations/Homo_sapiens.GRCh38.98.abinitio.gtf | awk '{if(!/^ *#/) print "chr"$0; else print;}' > annotations/Homo_sapiens.GRCh38.98.abinitio.ucsc.gtf`)

```{r, echo=T}
peaks <- Read10X_h5("~/my_data/filtered_peak_bc_matrix.h5")
chromosomes = c(1:22, "X", "Y")
chromosomes = paste("chr", chromosomes, sep="")
annotation.file <- "~/annotations/Homo_sapiens.GRCh38.86.gtf"
```

I rewrite the Seurat function because it becomes more portable between gene annotations.

```{r}
# Resize GenomicRanges upstream and or downstream
# from https://support.bioconductor.org/p/78652/
#
Extend <- function(x, upstream = 0, downstream = 0) {
  if (any(GenomicRanges::strand(x = x) == "*")) {
    warning("'*' ranges were treated as '+'")
  }
  on_plus <- GenomicRanges::strand(x = x) == "+" | GenomicRanges::strand(x = x) == "*"
  new_start <- GenomicRanges::start(x = x) - ifelse(test = on_plus, yes = upstream, no = downstream)
  new_end <- GenomicRanges::end(x = x) + ifelse(test = on_plus, yes = downstream, no = upstream)
  IRanges::ranges(x = x) <- IRanges::IRanges(start = new_start, end = new_end)
  x <- GenomicRanges::trim(x = x)
  return(x)
}

myCreateGeneActivityMatrix <- function( peak.matrix,
  annotation.file,
  seq.levels = c(1:22, "X", "Y"),
  include.body = TRUE,
  upstream = 2000,
  downstream = 0,
  verbose = TRUE){
    
  peak.df <- rownames(x = peak.matrix)
  peak.df <- do.call(what = rbind, args = strsplit(x = gsub(peak.df, pattern = ":", replacement = "-"), split = "-"))
  peak.df <- as.data.frame(x = peak.df)
  colnames(x = peak.df) <- c("chromosome", 'start', 'end')
  peaks.gr <- GenomicRanges::makeGRangesFromDataFrame(df = peak.df)
  
  # if any peaks start at 0, change to 1
  # otherwise GenomicRanges::distanceToNearest will not work 
  BiocGenerics::start(peaks.gr[BiocGenerics::start(peaks.gr) == 0, ]) <- 1

  # get annotation file, select genes
  gtf <- rtracklayer::import(con = annotation.file)

  # change seqlevelsStyle if not the same
  if (!any(GenomeInfoDb::seqlevelsStyle(x = gtf) == GenomeInfoDb::seqlevelsStyle(x = peaks.gr))) {
    GenomeInfoDb::seqlevelsStyle(gtf) <- GenomeInfoDb::seqlevelsStyle(peaks.gr)
  }
  gtf.genes <- gtf[gtf$type == 'gene']
  gtf <- GenomeInfoDb::keepSeqlevels(x = gtf, value = seq.levels, pruning.mode = 'coarse') ## I basically just swiched position of this line

  # Extend definition up/downstream
  if (include.body) {
    gtf.body_prom <- Extend(x = gtf.genes, upstream = upstream, downstream = downstream)
  } else {
    gtf.body_prom <- SummarizedExperiment::promoters(x = gtf.genes, upstream = upstream, downstream = downstream)
  }
  gene.distances <- GenomicRanges::distanceToNearest(x = peaks.gr, subject = gtf.body_prom)
  keep.overlaps <- gene.distances[rtracklayer::mcols(x = gene.distances)$distance == 0]
  peak.ids <- peaks.gr[S4Vectors::queryHits(x = keep.overlaps)]
  gene.ids <- gtf.genes[S4Vectors::subjectHits(x = keep.overlaps)]
  
  # Some GTF rows will not have gene_name attribute
  # Replace it by gene_id attribute
  gene.ids$gene_name[is.na(gene.ids$gene_name)] <- gene.ids$gene_id[is.na(gene.ids$gene_name)]
  
  peak.ids$gene.name <- gene.ids$gene_name
  peak.ids <- as.data.frame(x = peak.ids)
  peak.ids$peak <- rownames(peak.matrix)[S4Vectors::queryHits(x = keep.overlaps)]
  annotations <- peak.ids[, c('peak', 'gene.name')]
  colnames(x = annotations) <- c('feature', 'new_feature')

  # collapse into expression matrix
  peak.matrix <- as(object = peak.matrix, Class = 'matrix')
  all.features <- unique(x = annotations$new_feature)

  if (nbrOfWorkers() > 1) {
    mysapply <- future_sapply
  } else {
    mysapply <- ifelse(test = verbose, yes = pbsapply, no = sapply)
  }
  newmat <- mysapply(X = 1:length(x = all.features), FUN = function(x){
    features.use <- annotations[annotations$new_feature == all.features[[x]], ]$feature
    submat <- peak.matrix[features.use, ]
    if (length(x = features.use) > 1) {
      return(Matrix::colSums(x = submat))
    } else {
      return(submat)
    }
  })
  newmat <- t(x = newmat)
  rownames(x = newmat) <- all.features
  colnames(x = newmat) <- colnames(x = peak.matrix)
  return(as(object = newmat, Class = 'dgCMatrix'))
  }
```

```{r}
countMat <- myCreateGeneActivityMatrix(peak.matrix = peaks, annotation.file = annotation.file, seq.levels = chromosomes, upstream = 2000)

saveRDS(object = countMat, "~/my_data/cellranger-atac110_count_30439_WSSS8038360_GRCh38-1_1_0.geneActivity.RDS")
```

### Setup Seurat object

```{r}
thymus.atac <- CreateSeuratObject(counts = peaks, assay = "ATAC", project = "Thymus_ATAC")
thymus.atac[["ACTIVITY"]] <- CreateAssayObject(counts = countMat)
```

### Filter cells with little reads
```{r}
data.frame(nCount = thymus.atac$nCount_ATAC) %>%
  ggplot(aes(nCount)) + geom_histogram(bins=500) +
  geom_vline(xintercept = 4000, color="red")

thymus.atac <- subset(thymus.atac, subset = nCount_ATAC > 4000)
```

```{r}
data.frame(count_activity=thymus.atac$nCount_ACTIVITY, nFeature=thymus.atac$nFeature_ACTIVITY) %>%
  ggplot(aes(count_activity)) +
  geom_histogram(bins=200) +
  ggtitle("Total counts per cell")

data.frame(count_activity=thymus.atac$nCount_ACTIVITY, nFeature=thymus.atac$nFeature_ACTIVITY) %>%
  ggplot(aes(nFeature)) +
  geom_histogram(bins=200) +
  ggtitle("Total covered genes per cell")


```


### Data Preprocessing

```{r}
DefaultAssay(thymus.atac) <- "ACTIVITY"
# thymus.atac <- FindVariableFeatures(thymus.atac)
thymus.atac <- NormalizeData(thymus.atac, normalization.method = "LogNormalize", scale.factor = 1000)
thymus.atac <- ScaleData(thymus.atac)
``` 

<!-- ```{r, fig.height=10, fig.width=10} -->
<!-- smp_var_genes <- sample(VariableFeatures(thymus.atac), 10) -->
<!-- thymus.atac@assays$ACTIVITY@scale.data[c("CD4"),] %>% -->
<!--   as.matrix() %>% -->
<!--   melt(varnames=c("gene", "cell")) %>% -->
<!--   ggplot(aes(value)) + -->
<!--   geom_histogram(bins=100) + -->
<!--   facet_wrap(gene~.)  -->
<!-- ``` -->


## Viz UMAP
Testing different feature selection methods (coverage VS variability)
```{r, message=FALSE}
DefaultAssay(thymus.atac) <- "ACTIVITY"

## HVGs
thymus.atac.vargenes <- FindVariableFeatures(thymus.atac)

## Highly covered genes
thymus.atac.covgenes <- thymus.atac
VariableFeatures(thymus.atac.covgenes) <- names(which(Matrix::rowSums(thymus.atac.covgenes) > 100))

## Intersection
thymus.atac.varcovgenes <- thymus.atac
thymus.atac.varcovgenes <- FindVariableFeatures(thymus.atac.varcovgenes)
VariableFeatures(thymus.atac.varcovgenes) <- intersect(names(which(Matrix::rowSums(thymus.atac.covgenes) > 100)), VariableFeatures(thymus.atac.varcovgenes)) 
```

Coverage of HVGs
```{r}
gene.coverage <- Matrix::rowSums(thymus.atac@assays$ACTIVITY)
gene.meanAcc <- Matrix::rowMeans(thymus.atac@assays$ACTIVITY)

data.frame(cov=gene.coverage, gene=names(gene.coverage), meanAcc=gene.meanAcc) %>%
  mutate(var=ifelse(gene %in% VariableFeatures(thymus.atac.vargenes), "HVG", "Non-HVG")) %>%
  # mutate(var=ifelse(gene %in% VariableFeatures(thymus.atac.varcovgenes), "HVG", "Non-HVG")) %>%
  ggplot(aes(fill=var, cov)) + 
  geom_histogram(bins=50) +
  facet_wrap(var~., scales="free_y", nrow=2, ncol=1)

```

```{r, echo=T}
## Run embedding
embed <- function(thymus.atac){
  thymus.atac <- RunPCA(thymus.atac, assay = "ACTIVITY", verbose = F)
  thymus.atac <- RunUMAP(thymus.atac, reduction="pca", dims=1:50)
  return(thymus.atac)  
}

obj.list <- list(high.var=thymus.atac.vargenes, high.cov=thymus.atac.covgenes, high.var.cov = thymus.atac.varcovgenes)
obj.list.umap <- map(obj.list, ~ embed(.x))
imap(obj.list.umap, ~ DimPlot(.x, reduction='umap') + ggtitle(.y))
```

### Accessibility of marker genes 
```{r}
marker_genes <- c('CD4','CD8B','IFNG','KLRB1','AIRE','RAG1')
FeaturePlot(obj.list.umap$high.var, reduction = "umap", features = marker_genes)
```
```{r}
geneOI <- 'TOP2A'
FeaturePlot(obj.list.umap$high.var, reduction = "umap", features = geneOI)
```

### Clustering 
```{r}
thymus.atac.vargenes <- FindNeighbors(thymus.atac.vargenes)
thymus.atac.vargenes <- FindClusters(thymus.atac.vargenes)

```
```{r}
DimPlot(thymus.atac.vargenes, reduction = "umap") +
  ggtitle("louvain")
```



## Visualize UMAP on peaks (not gene activity)
Using Latent Semantic Indexing (LSI) as proposed by Cusanovich et al. 2015.
```{r, echo=TRUE}
DefaultAssay(thymus.atac) <- "ATAC"
VariableFeatures(thymus.atac) <- names(which(Matrix::rowSums(thymus.atac) > 100))
thymus.atac <- RunLSI(thymus.atac, n = 50, scale.max = NULL)
thymus.atac <- RunUMAP(thymus.atac, reduction = "lsi", dims = 1:50)

DimPlot(thymus.atac, reduction = "umap")
```

## Save gene activity matrix
```{r}
DropletUtils::write10xCounts(path = "~/my_data/cellranger-atac110_count_30439_WSSS8038360_GRCh38-1_1_0.geneActivity.hdf5",type = "HDF5", thymus.atac@assays$ACTIVITY@counts,
               gene.symbol=rownames(thymus.atac@assays$ACTIVITY@counts), barcodes=colnames(thymus.atac@assays$ACTIVITY@counts))
```

