---
title: "Thymus ATAC-seq preprocessing and EDA"
output: html_notebook
---

```{r}
suppressPackageStartupMessages({
library(Seurat)
library(tidyverse)
library(SingleCellExperiment)
library(scater)
library(pbapply)
library(future)
library(reshape2)
library(ggpubr)
library(UpSetR)
})

source("~/multiOmic_benchmark/selectFeatures.R")
```

Load F74 datasets (preprocessed as in `makeSCElist.r`). Includes raw and processed ATAC-seq and RNA-seq dataset. ATAC-seq data was reduced to gene level features as in Seurat pipeline. 

```{r}
f74.sce.list <- readRDS("~/my_data/integrated_thymus/F74_SCElist_20191017.RDS")

f74.atac <- f74.sce.list$ATAC
f74.rna <- f74.sce.list$RNA
```

## Feature selection
Compare betwen HVGs from ATAC and HVGs from RNA (i.e. genes used for cell type annotation)
```{r, message=FALSE, warning=FALSE}
n_features = 4000
hvg.rna <- HVG_Seurat(f74.sce.list[["RNA"]], nfeatures = n_features) %>% {.[. %in% rownames(f74.sce.list[["ATAC"]])]}
hvg.atac <- HVG_Seurat(f74.atac, nfeatures = length(hvg.rna))

upset(fromList(list(hvg.rna=hvg.rna, hvg.atac=hvg.atac)))
```
<!-- Check gene coverage -->
<!-- ```{r, fig.width=10, fig.height=6, echo=FALSE, eval=FALSE, message=FALSE, warning=FALSE} -->
<!-- gene.coverage <- Matrix::rowSums(counts(f74.atac)) -->

<!-- hvg.atac.pl <- data.frame(cov=gene.coverage, gene=names(gene.coverage)) %>% -->
<!--   dplyr::mutate(var=ifelse(gene %in% hvg.atac, "HVG-ATAC", "Non-HVG")) %>% -->
<!--   # mutate(var=ifelse(gene %in% VariableFeatures(thymus.atac.varcovgenes), "HVG", "Non-HVG")) %>% -->
<!--   ggplot(aes(fill=var, cov)) +  -->
<!--   geom_histogram(bins=50) + -->
<!--   facet_wrap(var~., scales="free_y", nrow=2, ncol=1) -->

<!-- hvg.rna.pl <- data.frame(cov=gene.coverage, gene=names(gene.coverage)) %>% -->
<!--   dplyr::mutate(var=ifelse(gene %in% hvg.rna, "HVG-RNA", "Non-HVG")) %>% -->
<!--   # mutate(var=ifelse(gene %in% VariableFeatures(thymus.atac.varcovgenes), "HVG", "Non-HVG")) %>% -->
<!--   ggplot(aes(fill=var, cov)) +  -->
<!--   geom_histogram(bins=50) + -->
<!--   facet_wrap(var~., scales="free_y", nrow=2, ncol=1) -->

<!-- ggarrange(hvg.atac.pl, hvg.rna.pl, common.legend = TRUE) -->
<!-- ``` -->

## Dimensionality Reduction
```{r, message=FALSE, warning=FALSE}
f74.atac.seu <- as.Seurat(f74.atac, verbose=FALSE)
f74.atac.seu <- ScaleData(f74.atac.seu, do.center = TRUE, verbose=FALSE)
f74.atac.seu <- RunPCA(f74.atac.seu, features = hvg.rna, verbose=FALSE) 
f74.atac.seu <- RunUMAP(f74.atac.seu, reduction = "pca", dims=1:30, verbose=FALSE) 

DimPlot(f74.atac.seu, reduction = "umap", group.by = NULL) + ggtitle("scRNA-seq HVGs") 
```

<!-- ATAC hvgs -->
<!-- ```{r} -->
<!-- f74.atac.seu.atachvg <- as.Seurat(f74.atac) -->
<!-- f74.atac.seu.atachvg <- ScaleData(f74.atac.seu.atachvg) -->
<!-- f74.atac.seu.atachvg <- RunPCA(f74.atac.seu.atachvg, features = hvg.atac)  -->
<!-- f74.atac.seu.atachvg <- RunUMAP(f74.atac.seu.atachvg, reduction = "pca", dims=1:30)  -->

<!-- DimPlot(f74.atac.seu.atachvg, reduction = "umap") + ggtitle("scATAC-seq HVGs")  -->
<!-- ``` -->

## Clustering 
```{r}
f74.atac.seu <- FindNeighbors(f74.atac.seu, verbose = F)
f74.atac.seu <- FindClusters(f74.atac.seu, algorithm = 4, verbose = F)

DimPlot(f74.atac.seu, reduction = "umap", group.by = 'ident') + ggtitle("leiden") +
  ggsave("~/multiOmic_benchmark/output/ATAC_EDA/ATAC_hvg_clustering.jpeg", height = 6, width=8)
```
<!-- ncd -->
<!-- ```{r} -->
<!-- f74.atac.seu.atachvg <- FindNeighbors(f74.atac.seu.atachvg) -->
<!-- f74.atac.seu.atachvg <- FindClusters(f74.atac.seu.atachvg, algorithm = 4) -->

<!-- DimPlot(f74.atac.seu.atachvg, reduction = "umap") + ggtitle("leiden") -->
<!-- ``` -->

## Accessibility of thymic markers
Provided by Cecilia
```{r, fig.width=10, fig.height=10, message=FALSE, warning=FALSE}
FeaturePlot(f74.atac.seu, features = c('PTPRC','CD4','CD8A','CD8B','CD79A','FOXN1','EPCAM','PDGFRA','GNG4', 'FOXP3','RAG1','RAG2','PTCRA','IL7R','NKG7','CCR7')) +
  ggsave("~/output/ATAC_EDA/ATAC_hvg_markers.jpeg", height = 15, width=15)
```

## Co-accessibility of markers

Genes that are distinctive of the same clusters will tend to be co-expressed in the same cells. Hence, to test if marker genes show patterning also in accessibility, I calculate correlation of accessibility profiles of thymus development marker genes and compare co-accessibility with their correlation in gene expression. 

```{r}
markers <- c('PTPRC','CD4','CD8A','CD8B','CD79A','FOXN1','EPCAM','PDGFRA','GNG4',
             'FOXP3','RAG1','RAG2','PTCRA','IL7R','NKG7','CCR7')
markers <- intersect(markers, rownames(f74.atac.seu@assays$RNA@counts))

## Calculate co-accessibility of marker genes
marker.cormat.atac <- 
  f74.atac.seu@assays$RNA@scale.data[markers,] %>%
  as.matrix() %>%
  t() %>%
  cor(use="pairwise.complete.obs")

## Same on RNA
f74.rna.seu <- as.Seurat(f74.rna)
f74.rna.seu <- ScaleData(f74.rna.seu)
marker.cormat.rna <- 
  f74.rna.seu@assays$RNA@scale.data[markers,] %>%
  as.matrix() %>%
  t() %>%
  cor(use="pairwise.complete.obs")
```

We can see that the co-accessibility signal is much weaker than the co-expression signal, but these are significantly correlated.

```{r}
png("~/multiOmic_benchmark/output/ATAC_EDA/ATAC_hvg_coaccessibility.png", height=500, width=500)
pheatmap::pheatmap(marker.cormat.atac, main="co-accessibility", cellwidth = 20, cellheight = 20)
dev.off()
png("~/multiOmic_benchmark/output/ATAC_EDA/ATAC_hvg_coexpression.png", height=500, width=500)
pheatmap::pheatmap(marker.cormat.rna, main="co-expression", cellwidth = 20, cellheight = 20)
dev.off()
```
```{r, warning=FALSE, message=FALSE}
long.cormat.atac <- 
  melt(marker.cormat.atac, value.name = "cor.accessibility") %>% 
  distinct(cor.accessibility, .keep_all = T) %>%
  unite("pair", c("Var1", "Var2"))

long.cormat.rna <- 
  melt(marker.cormat.rna, value.name = "cor.gex") %>% 
  distinct(cor.gex, .keep_all = T) %>%
  unite("pair", c("Var1", "Var2"))


full_join(long.cormat.atac, long.cormat.rna) %>%
  dplyr::filter(cor.accessibility!=1) %>%
  ggplot(aes(cor.accessibility, cor.gex)) +
  geom_point(alpha=0.5) +
  geom_vline(linetype=2, xintercept = 0) +
  geom_hline(linetype=2, yintercept = 0) +
  stat_cor() +
  ggtitle("Thymus marker genes") +
  ggsave("~/multiOmic_benchmark/output/ATAC_EDA/ATAC_hvg_markersCoexpression.png", height = 5, width=5)

```

<!-- Visualize some of the strong co-accessible markers -->
<!-- ```{r, fig.width=10, fig.height=5} -->
<!-- p1 <- FeatureScatter(f74.atac.seu, feature1 = "RAG1", feature2 = "RAG2", slot = "scale.data", smooth=TRUE) -->
<!-- p2 <- FeatureScatter(f74.rna.seu, feature1 = "RAG1", feature2 = "RAG2", slot = "scale.data", smooth=TRUE) -->

<!-- CombinePlots(plots=list(p1, p2)) -->
<!-- ``` -->

Control w random genes 
```{r, warning=FALSE, message=FALSE}

random <- sample(hvg.rna, size = 100)
random.cormat.atac <- 
  f74.atac.seu@assays$RNA@scale.data[random,] %>%
  as.matrix() %>%
  t() %>%
  cor(use="pairwise.complete.obs")

## Same on RNA
random.cormat.rna <- 
  f74.rna.seu@assays$RNA@scale.data[random,] %>%
  as.matrix() %>%
  t() %>%
  cor(use="pairwise.complete.obs")

long.random.cormat.atac <- 
  melt(random.cormat.atac, value.name = "cor.accessibility") %>% 
  distinct(cor.accessibility, .keep_all = T) %>%
  unite("pair", c("Var1", "Var2"))

long.random.cormat.rna <- 
  melt(random.cormat.rna, value.name = "cor.gex") %>% 
  distinct(cor.gex, .keep_all = T) %>%
  unite("pair", c("Var1", "Var2"))


full_join(long.random.cormat.atac, long.random.cormat.rna) %>%
  dplyr::filter(cor.accessibility!=1) %>%
  ggplot(aes(cor.accessibility, cor.gex)) +
  geom_point(alpha=0.5) +
  geom_vline(linetype=2, xintercept = 0) +
  geom_hline(linetype=2, yintercept = 0) +
  stat_cor() +
  ggtitle("Random HV genes") +
  ggsave("~/output/ATAC_EDA/ATAC_hvg_randomCoexpression.jpeg", height = 8, width=8)



```

## Fibroblasts markers
PDGFRA is not in the ATAC gene activity matrix (not covered enough?), so I check for genes with high co-expression with PDGFRA in RNA.


```{r, fig.height=8, fig.width=8}
fb_markers <- c("PDGFRA")

pdgfra.x <- f74.rna.seu@assays$RNA@scale.data["PDGFRA",]
pdgfra.cor <- apply(f74.rna.seu@assays$RNA@scale.data[hvg.rna,], 1, function(x) cor(x=x, y=pdgfra.x, use="complete.obs"))

fb_markers_atac <- pdgfra.cor[which(pdgfra.cor > 0.7)]

FeaturePlot(f74.atac.seu, features = names(fb_markers_atac)) 

```

Compare with expression in RNA
```{r, fig.height=8, fig.width=8}
f74.rna.seu <- ScaleData(f74.rna.seu, features = hvg.rna)

f74.rna.seu <- RunPCA(f74.rna.seu, features = hvg.rna) 
f74.rna.seu <- RunUMAP(f74.rna.seu, reduction = "pca", dims=1:30) 

FeaturePlot(f74.rna.seu, features = c(names(fb_markers_atac), "PDGFRA")) +
  ggsave("~/Fb_markers_RNA.pdf", height = 10, width=10)
```
```{r, fig.width=15, fig.height=10}
Seurat::VlnPlot(f74.atac.seu, features = c("MMP2", "COL12A1", "PLAT", "NTRK2"), slot = "data", ncol = 2)  
```

## Add cell type annotation to RNA data 
First clusters only 
```{r}
f74.rna.seu <- FindNeighbors(f74.rna.seu, verbose = F)
f74.rna.seu <- FindClusters(f74.rna.seu, algorithm = 4, verbose = F)

DimPlot(f74.rna.seu, reduction = "umap", group.by = 'ident') + ggtitle("leiden") +
  ggsave("~/multiOmic_benchmark/output/ATAC_EDA/RNA_hvg_clustering.jpeg", height = 6, width=8)

```

```{r}
annotation.df <- read.csv("~/my_data/F74_RNA_obs.csv")

# annotation.df$X <- str_remove(annotation.df$X, "F74_1_") %>% str_c("_1")
annotation.df <- annotation.df %>%
  mutate(cell=str_remove(as.character(X), "F74_1_") %>% str_c(ifelse(batch==0,'_1', "_2"))) 

f74.rna.seu@meta.data <-
  f74.rna.seu@meta.data %>%
  rownames_to_column("cell") %>%
  left_join(annotation.df, by="cell")

f74.rna.seu@meta.data %<>%
  column_to_rownames("cell")

ggplotly(DimPlot(f74.rna.seu, dims=1:2, group.by = "annotation", reduction = "umap"))
```

```{r}
FeaturePlot(f74.rna.seu, features = "HLA-DRA")
```


## Output clusters & UMAP coordinates for Daniel
```{r}
clusters_df <- f74.atac.seu@meta.data %>%
  as_tibble(rownames="cell") %>%
  select(cell, seurat_clusters)

umap_df <- 
  f74.atac.seu@reductions$umap@cell.embeddings %>%
  as_tibble(rownames="cell")

full_join(clusters_df, umap_df) %>%
  write_csv("~/my_data/integrated_thymus/F74_ATAC_clustersUMAP_20191021.csv")
```

