---
title: "Explore Gmat"
output: html_notebook
---

```{r}
library(tidyverse)
library(SingleCellExperiment)
library(Seurat)

## Make output directory
outdir <- "~/multiOmic_benchmark/report/output/20191125_gmat_EDA/"
ifelse(!dir.exists(outdir), dir.create(outdir), FALSE)
```

Explore distributions and statistics of gmat and gene expression, in order to instruct feature selection for integration

```{r}
rna.sce <- readRDS("~/my_data/F74_RNA_seurat_processed.RDS")
atac.sce <- readRDS("~/my_data/F74_ATAC_snapAtac_processed.RDS")

## Load gene model
transcripts.gr = rtracklayer::import("~/annotations/Homo_sapiens.GRCh38.93.filtered.gtf")
colnames(transcripts.gr@elementMetadata) <- str_replace(colnames(transcripts.gr@elementMetadata), "gene_name", "name")

genes.gr <- unlist(range(split(transcripts.gr, ~ name)))  ## From transcripts to genes
genes.gr$name <- names(genes.gr)
```

Filter genes with zero variance
```{r}
rna.gene.var <- as.matrix(counts(rna.sce)) %>% rowVars()
atac.gene.var <- as.matrix(counts(atac.sce)) %>% rowVars()

rna.sce <- rna.sce[which(rna.gene.var > 0),]
atac.sce <- atac.sce[which(atac.gene.var > 0),]

rna.sce; atac.sce
```


Is mean gene activity dependent on gene width?
```{r}
gene.width <- setNames(width(genes.gr), nm=genes.gr$name)

rna.gene.mean <- as.matrix(logcounts(rna.sce)) %>% rowMeans()
atac.gene.mean <- as.matrix(logcounts(atac.sce)) %>% rowMeans()
atac.gene.mean.scaled <- as.matrix(logcounts(atac.sce)) %>% t() %>% scale() %>% t() %>% rowMeans()

gene.mean.df <-
  full_join(
  data.frame(meanExpr=rna.gene.mean) %>% rownames_to_column("gene"),
  data.frame(meanAcc=atac.gene.mean) %>% rownames_to_column("gene")
) %>%
  full_join(
    data.frame(meanAcc.scaled=atac.gene.mean.scaled) %>% rownames_to_column("gene")
  ) %>%
  dplyr::mutate(width=gene.width[gene])

gene.mean.df %>%
  drop_na(width) %>%
  dplyr::filter(meanExpr < 50) %>%
  pivot_longer(cols=c("meanAcc.scaled", "meanAcc"), names_to = "tech", values_to = "mean.value") %>%
  ggplot(aes(  mean.value, width)) +
  facet_wrap(tech~., scales="free_x") +
  geom_bin2d(bins=100) +
  # scale_x_log10() +
  scale_fill_viridis_c()
```

### Embedding on highly-variable features from the gene matrix gives weird subclustering
Cells are embedded using Latent Semantic Indexing 
```{r}
atac.seu <- as.Seurat(atac.sce, counts = "counts")
atac.seu <- NormalizeData(atac.seu) 
atac.seu <- ScaleData(atac.seu) %>% RunLSI(n=30, reduction.name="lsi.activity") %>% RunUMAP(dims=1:30, reduction="lsi.activity", reduction.name="umap.activity")

pl1 <- DimPlot(atac.seu, reduction = "umap.activity", group.by="seurat_clusters", label = T) + ggtitle("gmat - HV features")
pl2 <- DimPlot(atac.seu, reduction = "UMAP", group.by="seurat_clusters", label=T) + ggtitle("bmat")
CombinePlots(list(pl1, pl2), legend = "top")
```

Same is true if I embed on gene accessibility of genes selected as highly variable in the RNA dataset

```{r}
rna.seu <- as.Seurat(rna.sce, counts="counts")
rna.seu <- FindVariableFeatures(rna.seu, 
                                # selection.method = "mvp", dispersion.cutoff = c(1,10),
                                #  mean.cutoff = c(0.5,7)
                                 )

VariableFeaturePlot(rna.seu)
```


```{r}
atac.seu.ref <- atac.seu
VariableFeatures(atac.seu.ref) <- reference.hvg
atac.seu.ref <- ScaleData(atac.seu.ref) %>% RunLSI(n=30, reduction.name="lsi.activity") %>% RunUMAP(dims=1:30, reduction="lsi.activity", reduction.name="umap.activity")

pl1 <- DimPlot(atac.seu.ref, reduction = "umap.activity", group.by="seurat_clusters", label = T) + ggtitle("gmat - HVG RNA")
CombinePlots(list(pl1, pl2), legend = "top")

```

<!-- Take marker genes for big cluster -->

<!-- ```{r, fig.height=12} -->
<!-- ## Pseudobulk x cluster -->
<!-- pseudobulk <- map(unique(as.numeric(atac.sce$seurat_clusters)), ~ counts(atac.sce)[,which(atac.sce$seurat_clusters==.x)] %>% rowSums()) %>% -->
<!--   setNames(unique(as.numeric(atac.sce$seurat_clusters))) -->


<!-- pseudobulk.mat <- reduce(pseudobulk, cbind) -->
<!-- colnames(pseudobulk.mat) <- names(pseudobulk) -->

<!-- pseudobulk.mat <- t(scale(t(log1p(pseudobulk.mat)))) -->

<!-- pseudo.tcells <- pseudobulk.mat[,as.character(1:9)] -->
<!-- pseudo.means <- rowMeans(pseudo.tcells) -->
<!-- pseudo.vars <- rowSds(pseudo.tcells) -->

<!-- loess.var <- loess(pseudo.vars ~ log(pseudo.means +1)) -->

<!-- plot(pseudo.means, pseudo.vars) -->

<!-- FindMarkers(atac.seu, ident.1 = 1, group.by = "seurat_clusters") -->

<!-- pseudobulk.mat[order(pseudo.vars, decreasing = T)[1:2000],] %>% -->
<!--   pheatmap::pheatmap() -->
<!-- ``` -->
<!-- ```{r} -->
<!-- top_feats <- rownames(pseudobulk.mat)[order(pseudo.vars, decreasing = T)[1:2000] ] -->
<!-- gene.mean.df %>% -->
<!--   dplyr::mutate(hvg=ifelse(gene %in% top_feats, "reference HVG", NA)) %>% -->
<!--   ggplot(aes(meanAcc, width)) + -->
<!--   geom_point() + -->
<!--   geom_point(data=. %>% dplyr::filter(!is.na(hvg)), color="red") -->

<!-- ``` -->

<!-- ```{r} -->
<!-- atac.seu.top <- atac.seu -->
<!-- VariableFeatures(atac.seu.top) <- top_feats -->
<!-- atac.seu.top <- ScaleData(atac.seu.top) %>% RunLSI(n=30, reduction.name="lsi.activity") %>% RunUMAP(dims=1:30, reduction="lsi.activity", reduction.name="umap.activity") -->
<!-- # atac.seu <- RunPCA(atac.seu, reduction.name="pca.activity") %>% RunUMAP(dims=1:30, reduction="pca.activity", reduction.name="umap.pca.activity") -->

<!-- pl1 <- DimPlot(atac.seu.top, reduction = "umap.activity", group.by="seurat_clusters", label = T) -->
<!-- pl2 <- DimPlot(atac.seu.top, reduction = "UMAP", group.by="seurat_clusters", label=T) -->
<!-- CombinePlots(list(pl1, pl2), legend = "top") -->
<!-- ``` -->

## Subset to T cells cluster

```{r}
tcells.sce <- atac.sce[,which(as.numeric(atac.sce$seurat_clusters) %in% c(1:9))]
tcells.seu <- atac.seu[,which(as.numeric(atac.seu$seurat_clusters) %in% c(1:9))]

```
```{r}
atac.gene.frac.tcells <- as.matrix(logcounts(tcells.sce)) %>% apply(1, function(x) sum(x!=0))
atac.gene.mean.tcells <- as.matrix(logcounts(tcells.sce)) %>% rowMeans()
atac.gene.max.tcells <- as.matrix(logcounts(tcells.sce)) %>% rowMax()

tcells.gene.mean.df <-
  full_join(
  data.frame(fracAcc=atac.gene.frac.tcells) %>% rownames_to_column("gene"),
  data.frame(meanAcc=atac.gene.mean.tcells) %>% rownames_to_column("gene")
) %>%
  full_join(
    data.frame(maxAcc=atac.gene.max.tcells, gene=rownames(logcounts(tcells.sce)))
  ) %>%
  dplyr::mutate(width=gene.width[gene])

```

```{r}
# tcells.seu <- as.Seurat(atac.sce, counts="counts", assay="ACTIVITY")
tcells.seu <- FindVariableFeatures(tcells.seu, selection.method = "mvp", dispersion.cutoff = c(1,10),
                                 mean.cutoff = c(0.5,5)
                                 )

VariableFeaturePlot(tcells.seu)
```

```{r}
tcells.seu <- NormalizeData(tcells.seu)
tcells.seu <- ScaleData(tcells.seu)

tcells.seu <- RunLSI(tcells.seu, n=30, reduction.name="lsi.activity", scale.max = NULL) 
tcells.seu <- RunUMAP(tcells.seu, dims=1:30, reduction="lsi.activity", reduction.name="umap.activity")

pl1 <- DimPlot(tcells.seu, reduction = "umap.activity", group.by="seurat_clusters", label = T)
pl2 <- DimPlot(tcells.seu, reduction = "UMAP", group.by="seurat_clusters", label=T)

CombinePlots(list(pl1, pl2), legend = "top")
```


```{r, fig.width=15, fig.height=6}
width.bins <- cut(rank(gene.width), breaks = 10) %>% setNames(names(gene.width))
width.bins.samples <- split(names(width.bins), width.bins) %>% lapply(function(x) sample(x, 5))
genes.sample <- unlist(width.bins.samples)
genes.sample <- intersect(genes.sample, rownames(rna.sce))

atac.logdata <- as.matrix(counts(atac.sce))[genes.sample,]
rna.logdata <- as.matrix(counts(rna.sce))[genes.sample,]

atac.scaledata <- atac.logdata[genes.sample,] %>% log1p() %>% t() %>% scale(scale=T) %>% t()
rna.scaledata <- rna.logdata[genes.sample,] %>% log1p() %>% t() %>% scale(scale=T) %>% t()

long.gene.mat <- function(mat){
  mat %>%
  reshape2::melt(varnames=c("gene", "cell")) %>%
  mutate(gene=factor(gene, levels = unique(gene)))
}

long.atac.logdata <- long.gene.mat(ifelse(atac.logdata==0, 0, atac.logdata))  %>% mutate(tranform = "log", data="ATAC")
long.rna.logdata <- long.gene.mat(ifelse(rna.logdata==0, 0, rna.logdata)) %>% mutate(tranform = "log", data="RNA")
long.atac.scaledata <- long.gene.mat(ifelse(atac.logdata==0, 0, atac.scaledata)) %>% mutate(tranform = "scale", data="ATAC")
long.rna.scaledata <- long.gene.mat(ifelse(rna.logdata==0, 0, rna.scaledata)) %>% mutate(tranform = "scale", data="RNA")

bind_rows(long.rna.logdata, long.rna.scaledata, long.atac.logdata, long.atac.scaledata) %>%
  filter(tranform=="log") %>%
  # filter(value>0) %>%
  mutate(width=gene.width[gene]) %>%
  ggplot(aes(gene, value)) +
  # ggbeeswarm::geom_quasirandom(alpha=0.1) +
  geom_violin() +
  # geom_boxplot(width=0.2, outlier.alpha = 0) +
  geom_jitter(size=0.1) +
  facet_grid(.~data, scales = "free") +
  theme_bw(base_size = 16) +
  theme(axis.text.x = element_blank()) 
  
  


```
```{r}
atac.gene4 <- atac.logdata[15,]
rna.gene4 <- rna.logdata[15,]

hist(logcounts(rna.sce)[genes.sample[15],])
hist(logcounts(atac.sce)[genes.sample[15],])
```

<!-- ```{r, fig.width=15} -->
<!-- rna.logdata <- as.matrix(logcounts(rna.sce)) -->
<!-- width.bins <- cut(rank(gene.width), breaks = 10) %>% setNames(names(gene.width)) -->
<!-- width.bins.samples <- split(names(width.bins), width.bins) %>% lapply(function(x) sample(x, 10)) -->

<!-- rna.scaledata <- rna.logdata[which(unlist(width.bins.samples) %in% rownames(rna.logdata)),] %>% t() %>% scale() %>% t() -->


<!-- # unlist(width.bins.samples)[which(!unlist(width.bins.samples) %in% rownames(rna.scaledata))] -->
<!-- rna.scaledata %>% -->
<!--  reshape2::melt(varnames=c("gene", "cell")) %>% -->
<!--   mutate(width=gene.width[gene], mean=rna.gene.mean[gene]) %>% -->
<!--   arrange(width) %>% -->
<!--   mutate(gene=factor(gene, levels = unique(gene))) %>% -->
<!--   ggplot(aes(gene, value)) + -->
<!--   geom_boxplot() + -->
<!--   ggtitle("scaled gene expr") -->
<!-- rna.logdata[which(unlist(width.bins.samples) %in% rownames(rna.logdata)),] %>% -->
<!--  reshape2::melt(varnames=c("gene", "cell")) %>% -->
<!--   mutate(width=gene.width[gene], mean=rna.gene.mean[gene]) %>% -->
<!--   arrange(width) %>% -->
<!--   mutate(gene=factor(gene, levels = unique(gene))) %>% -->
<!--   ggplot(aes(gene, value)) + -->
<!--   geom_boxplot() + -->
<!--   ggtitle("gene expr") -->



<!-- ``` -->
```{r}
rna.scaledata <- rna.logdata[sample(1:20000, 1000),] %>% t() %>% scale() %>% t()

atac.no0 <- 
  atac.logdata %>%
  ifelse(.==0, NA, .)
atac.no0.scaled <- 
  atac.scaledata %>%
  ifelse(is.na(atac.no0), NA, .)
rna.no0 <- 
  rna.logdata %>%
  ifelse(.==0, NA, .)
rna.no0.scaled <- 
  rna.scaledata %>%
  ifelse(is.na(rna.no0[rownames(rna.scaledata),]), NA, .)

atac.frac0 <- atac.logdata %>% apply(1, function(x) sum(x==0))
rna.frac0 <- rna.logdata %>% apply(1, function(x) sum(x==0))
atac.gene.mean.no0 <- atac.no0 %>% rowMeans(na.rm = T)
atac.gene.mean.no0.scaled <- atac.no0.scaled %>% rowMeans(na.rm = T)

rna.gene.mean.no0 <- rna.no0 %>% rowMeans(na.rm = T)
rna.gene.mean.no0.scaled <- rna.no0.scaled %>% rowMeans(na.rm = T)

data.frame(meanAcc.no0=atac.gene.mean.no0, meanExpr.no0=rna.gene.mean.no0[names(atac.gene.mean.no0)],
           frac0Acc = atac.frac0, frac0Expr = rna.frac0[names(atac.gene.mean.no0)],
           meanAcc.no0.scaled=atac.gene.mean.no0.scaled) %>%
  ggplot(aes(meanAcc.no0, meanAcc.no0.scaled)) +
  geom_hex(bins=100) +
  xlim(3,10) 

data.frame(meanExpr.no0=rna.gene.mean.no0[rownames(rna.scaledata)], meanExpr.no0.scaled = rna.gene.mean.no0.scaled[rownames(rna.scaledata)],
           frac0Expr = rna.frac0[rownames(rna.scaledata)], gene=rownames(rna.scaledata)) %>%
  # mutate(width=gene.width[gene]) +
  drop_na() %>%
  ggplot(aes(meanExpr.no0, meanExpr.no0.scaled)) +
  geom_hex(bins=100) +
    xlim(3,10) 


```

### Making the gmat bimodal
I convert any value that is different from 0 to 1. I select for LSI genes that are accessible in more than 500 cells.

```{r, fig.width=7, fig.height=7}
tcells.bimod.sce <- tcells.sce

logcounts(tcells.bimod.sce) <- 
  logcounts(tcells.bimod.sce) %>%
  as.matrix() %>%
  {ifelse(.==0, 0,1)}

tcells.bimod.seu <- as.Seurat(tcells.bimod.sce)
VariableFeatures(tcells.bimod.seu ) <- names(which(Matrix::rowSums(tcells.bimod.seu) > 500))
tcells.bimod.seu <- RunLSI(tcells.bimod.seu, n=50, scale.max = NULL)
tcells.bimod.seu <- RunUMAP(tcells.bimod.seu, reduction = "lsi", dims = 1:50)

pl1 <- DimPlot(tcells.bimod.seu, reduction = "umap", group.by="seurat_clusters", label = T) + ggtitle("Gene accessibility (binary)")
pl2 <- DimPlot(tcells.seu, reduction = "umap.activity", group.by="seurat_clusters", label = T) + ggtitle("Gene accessibility (counts)")
pl3 <- DimPlot(tcells.seu, reduction = "UMAP", group.by="seurat_clusters", label=T) + ggtitle("Genomic bins")

CombinePlots(list(pl3, pl2, pl1) %>% map(~ .x + theme(axis.ticks = element_blank(), axis.text=element_blank())), legend = "top", nrow=1) +
  ggsave(paste0(outdir, "gmatCountsVSgmatBin.png"), width=10, height = 4)
```

Selecting features based on coverage still enables better segregation of clusters compared to the RNA HVGs.

```{r}
tcells.bimod.seu.ref <- as.Seurat(tcells.bimod.sce)
VariableFeatures(tcells.bimod.seu.ref ) <- VariableFeatures(rna.seu)

tcells.bimod.seu.ref <- RunLSI(tcells.bimod.seu.ref, n=50, scale.max = NULL)
tcells.bimod.seu.ref <- RunUMAP(tcells.bimod.seu.ref, reduction = "lsi", dims = 1:50)

pl.ref <- DimPlot(tcells.bimod.seu.ref, reduction = "umap", group.by="seurat_clusters", label = T) + ggtitle("gmat binary - RNA HVG")

CombinePlots(list(pl.ref,pl1), legend = "top")
```

### Feature selection based on number of cells with accessible site
Selecting genes that are covered in at least 10% of the cells (~ 8000 genes in this case) still maintains the structure from the bmat.
```{r}
hist(Matrix::rowSums(tcells.bimod.seu), breaks=100, xlab="# accessible cells", main="# accessible cells per gene")
```
```{r, fig.height=7, fig.width=7}
feature_coverage_umap <- function(frac_cells){
  tcells.bimod.seu <- as.Seurat(tcells.bimod.sce)
  VariableFeatures(tcells.bimod.seu ) <- names(which(Matrix::rowSums(tcells.bimod.seu) > frac_cells*ncol(tcells.bimod.sce)))
  tcells.bimod.seu <- RunLSI(tcells.bimod.seu, n=50, scale.max = NULL)
  tcells.bimod.seu <- RunUMAP(tcells.bimod.seu, reduction = "lsi", dims = 1:50)
  return(tcells.bimod.seu)
}


feat.cov.umaps <- map(c(0.05, 0.1, 0.25,0.5), ~ feature_coverage_umap(.x))

pl.list <- feat.cov.umaps %>%
  set_names(c(0.05, 0.1, 0.25,0.5)) %>%
  imap( ~ DimPlot(.x, reduction = "umap",group.by="seurat_clusters", label = T) + ggtitle(glue('Frac. of cell coverage ={.y}')))
CombinePlots(pl.list, legend = "top")
```


<!-- Save SingleCellExperiment list with bimodal gene activity matrix -->
<!-- ```{r} -->
<!-- makeBimGmat <- function(atac.sce){ -->
<!--   logcounts(atac.sce) <-  -->
<!--     logcounts(atac.sce) %>% -->
<!--     as.matrix() %>% -->
<!--     {ifelse(.==0, 0,1)} -->
<!--   return(atac.sce) -->
<!-- } -->

<!-- bim.atac.sce <- makeBimGmat(atac.sce) -->

<!-- saveRDS(bim.atac.sce, "~/my_data/F74_ATAC_snapAtac_processed_bgmat.RDS") -->
<!-- ``` -->














