---
title: "Snap2bmat - SnapATAC data processing"
output: html_notebook
---

```{r}
suppressPackageStartupMessages({
  library(SnapATAC)
  library(leiden)
  library(umap)
  library(GenomicRanges)
  library(SingleCellExperiment)
  library(zoo)
  library(tidyverse)
  library(cowplot)
  library(ggpubr)
})
```

This processing pipeline follows the [SnapATAC integration vignette](https://github.com/r3fang/SnapATAC/blob/master/examples/10X_PBMC_15K/README.md). 

### Load Snap file
First I have made the snap file from the cellranger `fragments.tsv` file. Script is `multiOmic_benchmark/preprocess/fragments2snap.sh`.

```{r}
snap.files <- "~/my_data/cellranger-atac110_count_30439_WSSS8038360_GRCh38-1_1_0.snap"
sample.names <- "F74"
barcode.files <- '~/my_data/singlecell.csv'

x.sp.ls <- list(createSnap(snap.files, sample.names))
names(x.sp.ls) = sample.names
x.sp.ls

```


### Barcode selection
Filtering based on number of reads per cell and ratio of fragments that are within promoters
```{r}
barcode.ls = lapply(seq(snap.files), function(i){
    barcodes = read.csv(
        barcode.files[i], 
        head=TRUE
    );
    # remove NO BAROCDE line
    barcodes = barcodes[2:nrow(barcodes),];
    barcodes$logUMI = log10(barcodes$passed_filters + 1);
    barcodes$promoter_ratio = (barcodes$promoter_region_fragments+1) / (barcodes$passed_filters + 1);
    barcodes
  })

plots = lapply(seq(snap.files), function(i){
    p1 = ggplot(
        barcode.ls[[i]], 
        aes(x=logUMI, y=promoter_ratio)) + 
        geom_point(size=0.3, col="grey") +
        theme_classic()	+
        ggtitle(sample.names[[i]]) +
        ylim(0, 1) + xlim(0, 6) + 
        labs(x = "log10(counts)", y="promoter ratio")
        p1
    })

## Select and viz cutoffs
cutoff.logUMI.low = 3.3
cutoff.logUMI.high = 4.8
cutoff.FRIP.low = 0.25
cutoff.FRIP.high = 0.65


plots[[1]] + 
  geom_vline(xintercept = c(cutoff.logUMI.low[1],cutoff.logUMI.high[1]), linetype=2) +
  geom_hline(yintercept = c(cutoff.FRIP.low[1],cutoff.FRIP.high[1]), linetype=2)
```
```{r}
barcode.ls = lapply(seq(snap.files), function(i){
  barcodes = barcode.ls[[i]];
  idx = which(
      barcodes$logUMI >= cutoff.logUMI.low[i] & 
      barcodes$logUMI <= cutoff.logUMI.high[i] & 
      barcodes$promoter_ratio >= cutoff.FRIP.low[i] &
      barcodes$promoter_ratio <= cutoff.FRIP.high[i]
  );
  barcodes[idx,]
});
x.sp.ls = lapply(seq(snap.files), function(i){
  barcodes = barcode.ls[[i]];
  x.sp = x.sp.ls[[i]];
  barcode.shared = intersect(x.sp@barcode, barcodes$barcode);
  x.sp = x.sp[match(barcode.shared, x.sp@barcode),];
  barcodes = barcodes[match(barcode.shared, barcodes$barcode),];
  x.sp@metaData = barcodes;
  x.sp
})
names(x.sp.ls) = sample.names;
x.sp = Reduce(snapRbind, x.sp.ls);
x.sp@metaData["sample"] = x.sp@sample;
x.sp
```

### Add cell-by-bin matrix
```{r}
x.sp = addBmatToSnap(x.sp, bin.size = 5000)
```

### Binarize matrix
Some items in the count matrix have abnormally high coverage perhaps due to the alignment errors. Therefore, we next remove top 0.1% items in the count matrix and then convert the remaining non-zero values to 1.
```{r}
x.sp = makeBinary(x.sp, mat="bmat")
x.sp
```

### Filter bins
Filter out bins overlapping with ENCODE blacklist

```{r}
black_list = read.table("~/annotations/hg38.blacklist.bed.gz")
black_list.gr = GRanges(
  black_list[,1], 
  IRanges(black_list[,2], black_list[,3])
);
idy = queryHits(
  findOverlaps(x.sp@feature, black_list.gr)
);
if(length(idy) > 0){
  x.sp = x.sp[,-idy, mat="bmat"];
};
x.sp
```

Exclude bad chromosomes
```{r}
chr.exclude = seqlevels(x.sp@feature)[grep("random|chrM", seqlevels(x.sp@feature))]
idy = grep(paste(chr.exclude, collapse="|"), x.sp@feature)
if(length(idy) > 0){
  x.sp = x.sp[,-idy, mat="bmat"]
}
x.sp
```

remove the top 5% bins that overlap with invariant features such as the house keeping gene promoters

```{r}
bin.cov = log10(Matrix::colSums(x.sp@bmat)+1)
# bin.cov = Matrix::colSums(x.sp@bmat)
hist(
  bin.cov[bin.cov > 0], 
  xlab="log10(bin cov)", 
  main="log10(Bin Cov)", 
  col="lightblue", 
  # xlim=c(0, 5),
  breaks=100
);
bin.cutoff = quantile(bin.cov[bin.cov > 0], 0.95)
idy = which(bin.cov <= bin.cutoff & bin.cov > 0)
x.sp = x.sp[, idy, mat="bmat"];
x.sp
```

Remove any cells of bin coverage less than 1,000. The rational behind this is that some cells may have high number of unique fragments but end up with low bin coverage after filtering. This step is optional but highly recommanded.
```{r}
idx = which(Matrix::rowSums(x.sp@bmat) > 1000);
x.sp = x.sp[idx,];
x.sp
```

## Dimensionality reduction with LSI
```{r}
bmat <- x.sp@bmat
bmat <- t(bmat)
rownames(bmat) <- x.sp@feature$name
meta <- x.sp@metaData %>% rownames_to_column() %>% select(- c(cell_id,rowname) ) %>% column_to_rownames("barcode")

bmat.seu <- CreateSeuratObject(counts = bmat, assay = "ATAC", min.cells = 100, meta.data = meta)
bmat.seu <- RunLSI(bmat.seu, n = 30)
```
 
<!-- We remove the first component as it is strongly correlated to depth (as seen (here)[http://andrewjohnhill.com/images/posts/2019-5-6-dimensionality-reduction-for-scatac-data/analysis.html]) -->
<!-- ```{r} -->
<!-- FeaturePlot(bmat.seu, reduction = "lsi", dims = 5:6, feature='nCount_ATAC') -->
<!-- ``` -->

```{r}
bmat.seu <- FindNeighbors(bmat.seu, reduction = "lsi", dims = 1:30)
bmat.seu <- FindClusters(bmat.seu, algorithm = 4)
bmat.seu <- RunUMAP(bmat.seu, reduction = "lsi", dims = 1:30)

DimPlot(bmat.seu, reduction = "umap", label = T)
```

## Make gene activity matrix (gmat)

Summing counts over gene body + 2kb upstream (promoter)

```{r}
transcripts.gr = rtracklayer::import("~/annotations/Homo_sapiens.GRCh38.93.filtered.gtf")
colnames(transcripts.gr@elementMetadata) <- str_replace(colnames(transcripts.gr@elementMetadata), "gene_name", "name")

genes.gr <- unlist(range(split(transcripts.gr, ~ name)))  ## From transcripts to genes
genes.gr$name <- names(genes.gr)

if (GenomeInfoDb::seqlevelsStyle(genes.gr) != GenomeInfoDb::seqlevelsStyle(x.sp@feature) ) {
  GenomeInfoDb::seqlevelsStyle(genes.gr) <- GenomeInfoDb::seqlevelsStyle(x.sp@feature)
}

## Expand ranges to include promoter
expandRange = function(x, upstream=2000, downstream=1000) {
  strand_is_minus = strand(x) == "-"
  on_plus = which(!strand_is_minus)
  on_minus = which(strand_is_minus)
  start(x)[on_plus] = start(x)[on_plus] - upstream
  start(x)[on_minus] = start(x)[on_minus] - downstream
  end(x)[on_plus] = end(x)[on_plus] + downstream
  end(x)[on_minus] = end(x)[on_minus] + upstream
  x
}
genes.gr <- expandRange(genes.gr, downstream = 0)

## Make gmat
x.sp = createGmatFromMat(
    obj=x.sp, 
    input.mat="bmat",
    genes=genes.gr,
    do.par=TRUE,
    num.cores=10
  )

gmat <- t(x.sp@gmat)
bmat.seu[["ACTIVITY"]] <- CreateAssayObject(counts=gmat)
```

### Save Gmat as SingleCellExperiment object for benchmark
```{r}
# thymus.atac.act <- t(x.sp@gmat)
# thymus.atac.act <- thymus.atac.act[names(which(table(rownames(thymus.atac.act)) == 1)),] # Remove duplicate rows
thymus.atac.sce <- as.SingleCellExperiment(bmat.seu, assay = "ACTIVITY")
SingleCellExperiment::logcounts(thymus.atac.sce) <- NULL

counts <- SingleCellExperiment::counts(thymus.atac.sce)
cpm <- apply(counts, 2, function(x) (x/sum(x))*1e+06)
cpm(thymus.atac.sce) <- as(object = cpm, Class = "dgCMatrix")

logcounts(thymus.atac.sce) <- log1p(cpm(thymus.atac.sce))
rownames(thymus.atac.sce) %<>% str_replace_all("_", "-")

saveRDS(thymus.atac.sce, "~/my_data/F74_ATAC_snapAtac_processed.RDS")
```

## Make binary gmat and save
```{r}
makeBinGmat <- function(atac.sce){
  logcounts(atac.sce) <- 
    logcounts(atac.sce) %>%
    as.matrix() %>%
    {ifelse(.==0, 0,1)}
  return(atac.sce)
}

bim.atac.sce <- makeBinGmat(thymus.atac.sce)

saveRDS(bim.atac.sce, "~/my_data/F74_ATAC_snapAtac_processed_bgmat.RDS")
```

Save SnapATAC object 
```{r}
saveRDS(x.sp, file = "~/my_data/cellranger-atac110_count_30439_WSSS8038360_GRCh38-1_1_0.snapATAC.RDS")
```


